{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "from functools import reduce\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "from scipy.stats.stats import pearsonr\n",
    "from scipy import stats\n",
    "from scipy.stats import ks_2samp\n",
    "from scipy.stats import entropy"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Datafiles\n",
    "There are 96 separate datafiles, each representing the processed output of an RNAseq experiment looking at mRNA levels in wild-type (48 samples) or SNF2 deletion (48 samples) yeast cells grown under standard condition."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Reading individual files with Pandas\n",
    "\n",
    "The individual files are in the directory \"data/rawdata/\" and are tab-delimeted text files. These are basically raw spreadsheets stored as text, with individual lines separated by a **'\\n'** which indicated 'new line', and columns within each line separated by a **'\\t'** or 'tab' character. (We will also deal with files where the column is separated by a comma."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Like any good program for working with spreadsheet data, Pandas knows how to read this kind of file if we tell it what to look for. To do this we use its [read_csv](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_csv.html) function, specifying first the filepath, and then a few details about this file.  \n",
    "\n",
    "**sep=\"\\t\"** tells Pandas that the data separator is a tab character (as opposed to other common separators like comma or space)  \n",
    "**header=None** tells Pandas that the first line of the file is not a header (this is what it typically expects)  \n",
    "**names = ..** because there is no header, we tell Pandas what to call the columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>Expression</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>15S_rRNA</td>\n",
       "      <td>29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>21S_rRNA</td>\n",
       "      <td>221</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HRA1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ICR1</td>\n",
       "      <td>70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>LSR1</td>\n",
       "      <td>314</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7126</th>\n",
       "      <td>no_feature</td>\n",
       "      <td>414503</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7127</th>\n",
       "      <td>ambiguous</td>\n",
       "      <td>1387186</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7128</th>\n",
       "      <td>too_low_aQual</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7129</th>\n",
       "      <td>not_aligned</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7130</th>\n",
       "      <td>alignment_not_unique</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7131 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      Gene  Expression\n",
       "0                 15S_rRNA          29\n",
       "1                 21S_rRNA         221\n",
       "2                     HRA1           0\n",
       "3                     ICR1          70\n",
       "4                     LSR1         314\n",
       "...                    ...         ...\n",
       "7126            no_feature      414503\n",
       "7127             ambiguous     1387186\n",
       "7128         too_low_aQual           0\n",
       "7129           not_aligned           0\n",
       "7130  alignment_not_unique           0\n",
       "\n",
       "[7131 rows x 2 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.read_csv(\"data/rawdata/WT_rep16_MID92_allLanes_tophat2.0.5.bam.gbgout\",sep=\"\\t\",header=None,names=[\"Gene\",\"Expression\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If we just run that command, Pandas loads in the file and displays it. But we want to store the data, so we assign the data to a variable. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "ge = pd.read_csv(\"data/rawdata/WT_rep16_MID92_allLanes_tophat2.0.5.bam.gbgout\",sep=\"\\t\",header=None,names=[\"Gene\",\"Expression\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we can look at the file in different ways. The **head(n)** command shows us the first n rows."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>Expression</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>15S_rRNA</td>\n",
       "      <td>29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>21S_rRNA</td>\n",
       "      <td>221</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HRA1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ICR1</td>\n",
       "      <td>70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>LSR1</td>\n",
       "      <td>314</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>NME1</td>\n",
       "      <td>33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>PWR1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Q0010</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Q0017</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Q0032</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Gene  Expression\n",
       "0  15S_rRNA          29\n",
       "1  21S_rRNA         221\n",
       "2      HRA1           0\n",
       "3      ICR1          70\n",
       "4      LSR1         314\n",
       "5      NME1          33\n",
       "6      PWR1           0\n",
       "7     Q0010           0\n",
       "8     Q0017           0\n",
       "9     Q0032           0"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ge.head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Similarly the **tail(n)** command shows us the last n rows."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>Expression</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>7121</th>\n",
       "      <td>tY(GUA)J2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7122</th>\n",
       "      <td>tY(GUA)M1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7123</th>\n",
       "      <td>tY(GUA)M2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7124</th>\n",
       "      <td>tY(GUA)O</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7125</th>\n",
       "      <td>tY(GUA)Q</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7126</th>\n",
       "      <td>no_feature</td>\n",
       "      <td>414503</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7127</th>\n",
       "      <td>ambiguous</td>\n",
       "      <td>1387186</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7128</th>\n",
       "      <td>too_low_aQual</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7129</th>\n",
       "      <td>not_aligned</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7130</th>\n",
       "      <td>alignment_not_unique</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      Gene  Expression\n",
       "7121             tY(GUA)J2           0\n",
       "7122             tY(GUA)M1           0\n",
       "7123             tY(GUA)M2           0\n",
       "7124              tY(GUA)O           0\n",
       "7125              tY(GUA)Q           0\n",
       "7126            no_feature      414503\n",
       "7127             ambiguous     1387186\n",
       "7128         too_low_aQual           0\n",
       "7129           not_aligned           0\n",
       "7130  alignment_not_unique           0"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ge.tail(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "You can already see one (typical) problem, which is that the file is not just data. It also how some summary statistics that we don't care about. So we have to filter the rows in some manner. Here we use a bit of data about gene nomenclature in yeast. \n",
    "\n",
    "Most of the genes have this naming format:\n",
    "\n",
    "Systematic names for nuclear-encoded ORFs begin with the letter 'Y' (for 'Yeast'); the second letter denotes the chromosome number ('A' is chr I, 'B' is chr II, etc.); the third letter is either 'L' or 'R' for left or right chromosome arm; next is a three digit number indicating the order of the ORFs on that arm of a chromosome starting from the centromere, irrespective of strand; finally, there is an additional letter indicating the strand, either 'W' for Watson (the strand with 5' end at the left telomere) or 'C' for Crick (the complement strand, 5' end is at the right telomere).\n",
    "\n",
    "You can read about the yeast gene naming system here - http://seq.yeastgenome.org/help/community/nomenclature-conventions."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can look for rows that match this style by using the built in filtering system in Pandas. This specific command asks Pandas to return all rows where the Gene column starts with the character \"Y\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>Expression</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>YAL001C</td>\n",
       "      <td>232</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>YAL002W</td>\n",
       "      <td>245</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>YAL003W</td>\n",
       "      <td>8073</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>YAL004W</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>YAL005C</td>\n",
       "      <td>7471</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6741</th>\n",
       "      <td>YPR201W</td>\n",
       "      <td>45</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6742</th>\n",
       "      <td>YPR202W</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6743</th>\n",
       "      <td>YPR203W</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6744</th>\n",
       "      <td>YPR204C-A</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6745</th>\n",
       "      <td>YPR204W</td>\n",
       "      <td>633</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6685 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           Gene  Expression\n",
       "61      YAL001C         232\n",
       "62      YAL002W         245\n",
       "63      YAL003W        8073\n",
       "64      YAL004W           0\n",
       "65      YAL005C        7471\n",
       "...         ...         ...\n",
       "6741    YPR201W          45\n",
       "6742    YPR202W          22\n",
       "6743    YPR203W          18\n",
       "6744  YPR204C-A           0\n",
       "6745    YPR204W         633\n",
       "\n",
       "[6685 rows x 2 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ge[ge.Gene.str.startswith(\"Y\")]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "That accounts for 6685 of the 7130 rows. But what about the rest? We can start by reversing that filter to ask for rows that do not start with \"Y\". That work is done here with the **~** symbol."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>Expression</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>15S_rRNA</td>\n",
       "      <td>29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>21S_rRNA</td>\n",
       "      <td>221</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HRA1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ICR1</td>\n",
       "      <td>70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>LSR1</td>\n",
       "      <td>314</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7126</th>\n",
       "      <td>no_feature</td>\n",
       "      <td>414503</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7127</th>\n",
       "      <td>ambiguous</td>\n",
       "      <td>1387186</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7128</th>\n",
       "      <td>too_low_aQual</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7129</th>\n",
       "      <td>not_aligned</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7130</th>\n",
       "      <td>alignment_not_unique</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>446 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      Gene  Expression\n",
       "0                 15S_rRNA          29\n",
       "1                 21S_rRNA         221\n",
       "2                     HRA1           0\n",
       "3                     ICR1          70\n",
       "4                     LSR1         314\n",
       "...                    ...         ...\n",
       "7126            no_feature      414503\n",
       "7127             ambiguous     1387186\n",
       "7128         too_low_aQual           0\n",
       "7129           not_aligned           0\n",
       "7130  alignment_not_unique           0\n",
       "\n",
       "[446 rows x 2 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ge[~ge.Gene.str.startswith(\"Y\")]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This is going to look a bit messy, but we can ask Pandas to return all of the Gene names for these remaining rows, just so we can get a sense of what we're dealing with."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['15S_rRNA', '21S_rRNA', 'HRA1', 'ICR1', 'LSR1', 'NME1', 'PWR1',\n",
       "       'Q0010', 'Q0017', 'Q0032', 'Q0045', 'Q0050', 'Q0055', 'Q0060',\n",
       "       'Q0065', 'Q0070', 'Q0075', 'Q0080', 'Q0085', 'Q0092', 'Q0105',\n",
       "       'Q0110', 'Q0115', 'Q0120', 'Q0130', 'Q0140', 'Q0142', 'Q0143',\n",
       "       'Q0144', 'Q0160', 'Q0182', 'Q0250', 'Q0255', 'Q0275', 'Q0297',\n",
       "       'RDN18-1', 'RDN18-2', 'RDN25-1', 'RDN25-2', 'RDN37-1', 'RDN37-2',\n",
       "       'RDN5-1', 'RDN5-2', 'RDN5-3', 'RDN5-4', 'RDN5-5', 'RDN5-6',\n",
       "       'RDN58-1', 'RDN58-2', 'RNA170', 'RPM1', 'RPR1', 'RUF20', 'RUF21',\n",
       "       'RUF22', 'RUF23', 'RUF5-1', 'RUF5-2', 'SCR1', 'SRG1', 'TLC1',\n",
       "       'snR10', 'snR11', 'snR128', 'snR13', 'snR14', 'snR161', 'snR17a',\n",
       "       'snR17b', 'snR18', 'snR189', 'snR19', 'snR190', 'snR191', 'snR24',\n",
       "       'snR3', 'snR30', 'snR31', 'snR32', 'snR33', 'snR34', 'snR35',\n",
       "       'snR36', 'snR37', 'snR38', 'snR39', 'snR39B', 'snR4', 'snR40',\n",
       "       'snR41', 'snR42', 'snR43', 'snR44', 'snR45', 'snR46', 'snR47',\n",
       "       'snR48', 'snR49', 'snR5', 'snR50', 'snR51', 'snR52', 'snR53',\n",
       "       'snR54', 'snR55', 'snR56', 'snR57', 'snR58', 'snR59', 'snR6',\n",
       "       'snR60', 'snR61', 'snR62', 'snR63', 'snR64', 'snR65', 'snR66',\n",
       "       'snR67', 'snR68', 'snR69', 'snR7-L', 'snR7-S', 'snR70', 'snR71',\n",
       "       'snR72', 'snR73', 'snR74', 'snR75', 'snR76', 'snR77', 'snR78',\n",
       "       'snR79', 'snR8', 'snR80', 'snR81', 'snR82', 'snR83', 'snR84',\n",
       "       'snR85', 'snR86', 'snR87', 'snR9', 'tA(AGC)D', 'tA(AGC)F',\n",
       "       'tA(AGC)G', 'tA(AGC)H', 'tA(AGC)J', 'tA(AGC)K1', 'tA(AGC)K2',\n",
       "       'tA(AGC)L', 'tA(AGC)M1', 'tA(AGC)M2', 'tA(AGC)P', 'tA(UGC)A',\n",
       "       'tA(UGC)E', 'tA(UGC)G', 'tA(UGC)L', 'tA(UGC)O', 'tA(UGC)Q',\n",
       "       'tC(GCA)B', 'tC(GCA)G', 'tC(GCA)P1', 'tC(GCA)P2', 'tC(GCA)Q',\n",
       "       'tD(GUC)B', 'tD(GUC)D', 'tD(GUC)G1', 'tD(GUC)G2', 'tD(GUC)I1',\n",
       "       'tD(GUC)I2', 'tD(GUC)J1', 'tD(GUC)J2', 'tD(GUC)J3', 'tD(GUC)J4',\n",
       "       'tD(GUC)K', 'tD(GUC)L1', 'tD(GUC)L2', 'tD(GUC)M', 'tD(GUC)N',\n",
       "       'tD(GUC)O', 'tD(GUC)Q', 'tE(CUC)D', 'tE(CUC)I', 'tE(UUC)B',\n",
       "       'tE(UUC)C', 'tE(UUC)E1', 'tE(UUC)E2', 'tE(UUC)E3', 'tE(UUC)G1',\n",
       "       'tE(UUC)G2', 'tE(UUC)G3', 'tE(UUC)I', 'tE(UUC)J', 'tE(UUC)K',\n",
       "       'tE(UUC)L', 'tE(UUC)M', 'tE(UUC)P', 'tE(UUC)Q', 'tF(GAA)B',\n",
       "       'tF(GAA)D', 'tF(GAA)F', 'tF(GAA)G', 'tF(GAA)H1', 'tF(GAA)H2',\n",
       "       'tF(GAA)M', 'tF(GAA)N', 'tF(GAA)P1', 'tF(GAA)P2', 'tF(GAA)Q',\n",
       "       'tG(CCC)D', 'tG(CCC)O', 'tG(GCC)B', 'tG(GCC)C', 'tG(GCC)D1',\n",
       "       'tG(GCC)D2', 'tG(GCC)E', 'tG(GCC)F1', 'tG(GCC)F2', 'tG(GCC)G1',\n",
       "       'tG(GCC)G2', 'tG(GCC)J1', 'tG(GCC)J2', 'tG(GCC)M', 'tG(GCC)O1',\n",
       "       'tG(GCC)O2', 'tG(GCC)P1', 'tG(GCC)P2', 'tG(UCC)G', 'tG(UCC)N',\n",
       "       'tG(UCC)O', 'tG(UCC)Q', 'tH(GUG)E1', 'tH(GUG)E2', 'tH(GUG)G1',\n",
       "       'tH(GUG)G2', 'tH(GUG)H', 'tH(GUG)K', 'tH(GUG)M', 'tH(GUG)Q',\n",
       "       'tI(AAU)B', 'tI(AAU)D', 'tI(AAU)E1', 'tI(AAU)E2', 'tI(AAU)G',\n",
       "       'tI(AAU)I1', 'tI(AAU)I2', 'tI(AAU)L1', 'tI(AAU)L2', 'tI(AAU)N1',\n",
       "       'tI(AAU)N2', 'tI(AAU)P1', 'tI(AAU)P2', 'tI(GAU)Q', 'tI(UAU)D',\n",
       "       'tI(UAU)L', 'tK(CUU)C', 'tK(CUU)D1', 'tK(CUU)D2', 'tK(CUU)E1',\n",
       "       'tK(CUU)E2', 'tK(CUU)F', 'tK(CUU)G1', 'tK(CUU)G2', 'tK(CUU)G3',\n",
       "       'tK(CUU)I', 'tK(CUU)J', 'tK(CUU)K', 'tK(CUU)M', 'tK(CUU)P',\n",
       "       'tK(UUU)D', 'tK(UUU)G1', 'tK(UUU)G2', 'tK(UUU)K', 'tK(UUU)L',\n",
       "       'tK(UUU)O', 'tK(UUU)P', 'tK(UUU)Q', 'tL(CAA)A', 'tL(CAA)C',\n",
       "       'tL(CAA)D', 'tL(CAA)G1', 'tL(CAA)G2', 'tL(CAA)G3', 'tL(CAA)K',\n",
       "       'tL(CAA)L', 'tL(CAA)M', 'tL(CAA)N', 'tL(GAG)G', 'tL(UAA)B1',\n",
       "       'tL(UAA)B2', 'tL(UAA)D', 'tL(UAA)J', 'tL(UAA)K', 'tL(UAA)L',\n",
       "       'tL(UAA)N', 'tL(UAA)Q', 'tL(UAG)J', 'tL(UAG)L1', 'tL(UAG)L2',\n",
       "       'tM(CAU)C', 'tM(CAU)D', 'tM(CAU)E', 'tM(CAU)J1', 'tM(CAU)J2',\n",
       "       'tM(CAU)J3', 'tM(CAU)M', 'tM(CAU)O1', 'tM(CAU)O2', 'tM(CAU)P',\n",
       "       'tM(CAU)Q1', 'tM(CAU)Q2', 'tN(GUU)C', 'tN(GUU)F', 'tN(GUU)G',\n",
       "       'tN(GUU)K', 'tN(GUU)L', 'tN(GUU)N1', 'tN(GUU)N2', 'tN(GUU)O1',\n",
       "       'tN(GUU)O2', 'tN(GUU)P', 'tN(GUU)Q', 'tP(AGG)C', 'tP(AGG)N',\n",
       "       'tP(UGG)A', 'tP(UGG)F', 'tP(UGG)H', 'tP(UGG)L', 'tP(UGG)M',\n",
       "       'tP(UGG)N1', 'tP(UGG)N2', 'tP(UGG)O1', 'tP(UGG)O2', 'tP(UGG)O3',\n",
       "       'tP(UGG)Q', 'tQ(CUG)M', 'tQ(UUG)B', 'tQ(UUG)C', 'tQ(UUG)D1',\n",
       "       'tQ(UUG)D2', 'tQ(UUG)D3', 'tQ(UUG)E1', 'tQ(UUG)E2', 'tQ(UUG)H',\n",
       "       'tQ(UUG)L', 'tQ(UUG)Q', 'tR(ACG)D', 'tR(ACG)E', 'tR(ACG)J',\n",
       "       'tR(ACG)K', 'tR(ACG)L', 'tR(ACG)O', 'tR(ACG)Q2', 'tR(CCG)L',\n",
       "       'tR(CCU)J', 'tR(UCU)B', 'tR(UCU)D', 'tR(UCU)E', 'tR(UCU)G1',\n",
       "       'tR(UCU)G2', 'tR(UCU)G3', 'tR(UCU)J1', 'tR(UCU)J2', 'tR(UCU)K',\n",
       "       'tR(UCU)M1', 'tR(UCU)M2', 'tR(UCU)Q1', 'tS(AGA)A', 'tS(AGA)B',\n",
       "       'tS(AGA)D1', 'tS(AGA)D2', 'tS(AGA)D3', 'tS(AGA)E', 'tS(AGA)G',\n",
       "       'tS(AGA)H', 'tS(AGA)J', 'tS(AGA)L', 'tS(AGA)M', 'tS(CGA)C',\n",
       "       'tS(GCU)F', 'tS(GCU)L', 'tS(GCU)O', 'tS(GCU)Q1', 'tS(UGA)E',\n",
       "       'tS(UGA)I', 'tS(UGA)P', 'tS(UGA)Q2', 'tT(AGU)B', 'tT(AGU)C',\n",
       "       'tT(AGU)D', 'tT(AGU)H', 'tT(AGU)I1', 'tT(AGU)I2', 'tT(AGU)J',\n",
       "       'tT(AGU)N1', 'tT(AGU)N2', 'tT(AGU)O1', 'tT(AGU)O2', 'tT(CGU)K',\n",
       "       'tT(UGU)G1', 'tT(UGU)G2', 'tT(UGU)H', 'tT(UGU)P', 'tT(UGU)Q1',\n",
       "       'tT(XXX)Q2', 'tV(AAC)E1', 'tV(AAC)E2', 'tV(AAC)G1', 'tV(AAC)G2',\n",
       "       'tV(AAC)G3', 'tV(AAC)H', 'tV(AAC)J', 'tV(AAC)K1', 'tV(AAC)K2',\n",
       "       'tV(AAC)L', 'tV(AAC)M1', 'tV(AAC)M2', 'tV(AAC)M3', 'tV(AAC)O',\n",
       "       'tV(CAC)D', 'tV(CAC)H', 'tV(UAC)B', 'tV(UAC)D', 'tV(UAC)Q',\n",
       "       'tW(CCA)G1', 'tW(CCA)G2', 'tW(CCA)J', 'tW(CCA)K', 'tW(CCA)M',\n",
       "       'tW(CCA)P', 'tW(UCA)Q', 'tX(XXX)D', 'tY(GUA)D', 'tY(GUA)F1',\n",
       "       'tY(GUA)F2', 'tY(GUA)J1', 'tY(GUA)J2', 'tY(GUA)M1', 'tY(GUA)M2',\n",
       "       'tY(GUA)O', 'tY(GUA)Q', 'no_feature', 'ambiguous', 'too_low_aQual',\n",
       "       'not_aligned', 'alignment_not_unique'], dtype=object)"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ge[~ge.Gene.str.startswith(\"Y\")].Gene.values"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "There are several types of things in here.  \n",
    "  \n",
    "**ribosomal RNAs** '15S_rRNA', '21S_rRNA'  \n",
    "**non-coding RNAs** 'HRA1', 'ICR1', 'LSR1', 'NME1', 'PWR1', etc..    \n",
    "**dubious open-reading frames** 'Q0010', 'Q0017', 'Q0032', 'Q0045', etc...  \n",
    "**snoRNAs** snR10', 'snR11', 'snR128', 'snR13', 'snR14', 'snR161', 'snR17a', etc...  \n",
    "**tRNAs** 'tA(AGC)D', 'tA(AGC)F','tA(AGC)G', 'tA(AGC)H', 'tA(AGC)J', 'tA(AGC)K1', etc...  \n",
    "**summary rows - not genes** 'no_feature', 'ambiguous', 'too_low_aQual','not_aligned', 'alignment_not_unique'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For many subsequent analyses, we'll stick to protein-coding genes. But for now, lets just get rid of the superfluous stuff."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "rows_to_drop = ['no_feature', 'ambiguous', 'too_low_aQual','not_aligned', 'alignment_not_unique']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "ge = ge[~ge.Gene.isin(rows_to_drop)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>Expression</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>15S_rRNA</td>\n",
       "      <td>29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>21S_rRNA</td>\n",
       "      <td>221</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HRA1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ICR1</td>\n",
       "      <td>70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>LSR1</td>\n",
       "      <td>314</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>NME1</td>\n",
       "      <td>33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>PWR1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Q0010</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Q0017</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Q0032</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Gene  Expression\n",
       "0  15S_rRNA          29\n",
       "1  21S_rRNA         221\n",
       "2      HRA1           0\n",
       "3      ICR1          70\n",
       "4      LSR1         314\n",
       "5      NME1          33\n",
       "6      PWR1           0\n",
       "7     Q0010           0\n",
       "8     Q0017           0\n",
       "9     Q0032           0"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ge.head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "So, we've got 96 files like this. For what we want to do, we want to combine these all into one giant table. Fortunately, pandas has a relative easy way to merge dataframes."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To illustrate this, first lets load two:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "ge1 = pd.read_csv(\"data/rawdata/WT_rep16_MID92_allLanes_tophat2.0.5.bam.gbgout\",sep=\"\\t\",header=None,names=[\"Gene\",\"Expression\"])\n",
    "ge1 = ge1[~ge1.Gene.isin(rows_to_drop)]\n",
    "ge2 = pd.read_csv(\"data/rawdata/Snf2_rep03_MID28_allLanes_tophat2.0.5.bam.gbgout\",sep=\"\\t\",header=None,names=[\"Gene\",\"Expression\"])\n",
    "ge2 = ge2[~ge2.Gene.isin(rows_to_drop)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>Expression</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>15S_rRNA</td>\n",
       "      <td>29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>21S_rRNA</td>\n",
       "      <td>221</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HRA1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ICR1</td>\n",
       "      <td>70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>LSR1</td>\n",
       "      <td>314</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>NME1</td>\n",
       "      <td>33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>PWR1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Q0010</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Q0017</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Q0032</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Gene  Expression\n",
       "0  15S_rRNA          29\n",
       "1  21S_rRNA         221\n",
       "2      HRA1           0\n",
       "3      ICR1          70\n",
       "4      LSR1         314\n",
       "5      NME1          33\n",
       "6      PWR1           0\n",
       "7     Q0010           0\n",
       "8     Q0017           0\n",
       "9     Q0032           0"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ge1.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>Expression</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>15S_rRNA</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>21S_rRNA</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HRA1</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ICR1</td>\n",
       "      <td>211</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>LSR1</td>\n",
       "      <td>159</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>NME1</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>PWR1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Q0010</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Q0017</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Q0032</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Gene  Expression\n",
       "0  15S_rRNA           5\n",
       "1  21S_rRNA          23\n",
       "2      HRA1           3\n",
       "3      ICR1         211\n",
       "4      LSR1         159\n",
       "5      NME1          17\n",
       "6      PWR1           0\n",
       "7     Q0010           0\n",
       "8     Q0017           0\n",
       "9     Q0032           0"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ge2.head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The merge syntax for merging two dataframes df1, df2 is pretty easy  \n",
    "\n",
    "**df1.merge(df2,on=\"Gene\",how=\"outer\")**\n",
    "\n",
    "In addition to the two dataframes, you have to specify what column of the dataframes you want to merge on. When pandas finds the same value in this column between two dataframes, it will combine them into one row. There are different ways this can be done - basically you can choose whether to use any value from either dataframe (how=\"outer\") only values found in both dataframes (how=\"inner\") only values in the first dataframe (how=\"left\") or in the second (how=\"right\")."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>Expression_x</th>\n",
       "      <th>Expression_y</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>15S_rRNA</td>\n",
       "      <td>29</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>21S_rRNA</td>\n",
       "      <td>221</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HRA1</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ICR1</td>\n",
       "      <td>70</td>\n",
       "      <td>211</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>LSR1</td>\n",
       "      <td>314</td>\n",
       "      <td>159</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7121</th>\n",
       "      <td>tY(GUA)J2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7122</th>\n",
       "      <td>tY(GUA)M1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7123</th>\n",
       "      <td>tY(GUA)M2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7124</th>\n",
       "      <td>tY(GUA)O</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7125</th>\n",
       "      <td>tY(GUA)Q</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7126 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           Gene  Expression_x  Expression_y\n",
       "0      15S_rRNA            29             5\n",
       "1      21S_rRNA           221            23\n",
       "2          HRA1             0             3\n",
       "3          ICR1            70           211\n",
       "4          LSR1           314           159\n",
       "...         ...           ...           ...\n",
       "7121  tY(GUA)J2             0             0\n",
       "7122  tY(GUA)M1             0             1\n",
       "7123  tY(GUA)M2             0             0\n",
       "7124   tY(GUA)O             0             0\n",
       "7125   tY(GUA)Q             0             0\n",
       "\n",
       "[7126 rows x 3 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ge1.merge(ge2,on=\"Gene\",how='outer')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The nice thing about merge is that it doesn't care what order the data are in. We can show this by sorting ge1 and ge2 on in different ways and remerge."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "ge1 = ge1.sort_values('Expression')\n",
    "ge2 = ge2.sort_values('Gene')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>Expression_x</th>\n",
       "      <th>Expression_y</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1549</th>\n",
       "      <td>15S_rRNA</td>\n",
       "      <td>29</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3614</th>\n",
       "      <td>21S_rRNA</td>\n",
       "      <td>221</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>216</th>\n",
       "      <td>HRA1</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1985</th>\n",
       "      <td>ICR1</td>\n",
       "      <td>70</td>\n",
       "      <td>211</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4389</th>\n",
       "      <td>LSR1</td>\n",
       "      <td>314</td>\n",
       "      <td>159</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>186</th>\n",
       "      <td>tY(GUA)J2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>172</th>\n",
       "      <td>tY(GUA)M1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>217</th>\n",
       "      <td>tY(GUA)M2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144</th>\n",
       "      <td>tY(GUA)O</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>tY(GUA)Q</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7126 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           Gene  Expression_x  Expression_y\n",
       "1549   15S_rRNA            29             5\n",
       "3614   21S_rRNA           221            23\n",
       "216        HRA1             0             3\n",
       "1985       ICR1            70           211\n",
       "4389       LSR1           314           159\n",
       "...         ...           ...           ...\n",
       "186   tY(GUA)J2             0             0\n",
       "172   tY(GUA)M1             0             1\n",
       "217   tY(GUA)M2             0             0\n",
       "144    tY(GUA)O             0             0\n",
       "0      tY(GUA)Q             0             0\n",
       "\n",
       "[7126 rows x 3 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ge1.merge(ge2,on=\"Gene\",how='outer').sort_values('Gene')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "How do we do this for 96 files?  \n",
    "\n",
    "First, let's get a list of all the files. To do this we use the "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Snf2_rep39_MID19_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep18_MID48_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep17_MID66_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep31_MID77_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep43_MID73_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep35_MID60_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep28_MID74_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep08_MID78_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep30_MID16_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep08_MID20_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep40_MID12_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep40_MID18_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep28_MID64_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep06_MID30_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep14_MID33_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep03_MID51_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep01_MID96_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep31_MID43_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep42_MID10_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep30_MID27_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep48_MID55_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep20_MID90_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep03_MID28_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep48_MID85_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep38_MID06_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep15_MID75_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep38_MID87_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep45_MID86_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep45_MID44_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep15_MID09_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep34_MID76_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep19_MID84_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep07_MID24_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep33_MID88_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep41_MID13_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep05_MID82_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep43_MID49_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep13_MID59_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep27_MID41_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep19_MID83_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep10_MID37_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep07_MID26_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep36_MID23_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep10_MID38_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep46_MID35_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep35_MID32_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep20_MID15_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep26_MID08_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep12_MID02_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep47_MID17_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep16_MID92_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep24_MID42_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep27_MID34_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep41_MID79_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep21_MID31_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep09_MID04_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep24_MID69_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep26_MID71_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep09_MID91_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep13_MID45_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep23_MID36_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep37_MID46_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep14_MID52_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep32_MID70_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep05_MID93_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep42_MID47_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep04_MID11_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep29_MID57_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep25_MID61_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep18_MID40_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep34_MID14_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep39_MID05_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep11_MID54_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep02_MID21_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep29_MID68_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep01_MID39_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep04_MID58_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep21_MID62_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep12_MID03_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep25_MID81_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep47_MID56_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep22_MID67_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep46_MID94_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep44_MID65_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep32_MID95_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep22_MID53_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep16_MID72_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep17_MID89_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep11_MID01_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep23_MID25_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep06_MID80_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep33_MID29_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep44_MID22_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep02_MID50_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep37_MID07_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep36_MID63_allLanes_tophat2.0.5.bam.gbgout']"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "os.listdir('data/rawdata')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Would be nicer if this were sorted"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Snf2_rep01_MID96_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep02_MID21_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep03_MID28_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep04_MID11_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep05_MID82_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep06_MID80_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep07_MID26_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep08_MID78_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep09_MID04_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep10_MID38_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep11_MID01_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep12_MID02_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep13_MID59_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep14_MID33_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep15_MID75_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep16_MID72_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep17_MID89_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep18_MID40_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep19_MID84_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep20_MID15_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep21_MID31_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep22_MID53_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep23_MID25_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep24_MID42_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep25_MID81_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep26_MID71_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep27_MID34_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep28_MID74_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep29_MID57_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep30_MID16_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep31_MID77_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep32_MID70_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep33_MID29_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep34_MID14_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep35_MID60_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep36_MID23_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep37_MID46_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep38_MID06_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep39_MID19_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep40_MID12_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep41_MID79_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep42_MID47_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep43_MID73_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep44_MID65_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep45_MID86_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep46_MID35_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep47_MID56_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep48_MID55_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep01_MID39_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep02_MID50_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep03_MID51_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep04_MID58_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep05_MID93_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep06_MID30_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep07_MID24_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep08_MID20_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep09_MID91_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep10_MID37_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep11_MID54_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep12_MID03_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep13_MID45_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep14_MID52_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep15_MID09_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep16_MID92_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep17_MID66_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep18_MID48_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep19_MID83_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep20_MID90_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep21_MID62_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep22_MID67_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep23_MID36_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep24_MID69_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep25_MID61_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep26_MID08_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep27_MID41_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep28_MID64_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep29_MID68_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep30_MID27_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep31_MID43_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep32_MID95_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep33_MID88_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep34_MID76_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep35_MID32_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep36_MID63_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep37_MID07_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep38_MID87_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep39_MID05_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep40_MID18_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep41_MID13_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep42_MID10_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep43_MID49_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep44_MID22_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep45_MID44_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep46_MID94_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep47_MID17_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep48_MID85_allLanes_tophat2.0.5.bam.gbgout']"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sorted(os.listdir('data/rawdata'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The files contain information about the experiment, which we want to extract. We'll do this using the 'split' command, which takes a string and splits it up based on some character in the string. Here we care about the first two parts of the file name, whether it's WT or mutant, and which replicate it is.  \n",
    "\n",
    "Let's illustrate what we're going to do with an example."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "filename = 'WT_rep25_MID61_allLanes_tophat2.0.5.bam.gbgout'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We want to split the string using the '_' character."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['WT', 'rep25', 'MID61', 'allLanes', 'tophat2.0.5.bam.gbgout']"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filename.split('_')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Exp type: WT\n",
      "Rep: rep25\n"
     ]
    }
   ],
   "source": [
    "filesplit = filename.split('_')\n",
    "exp_type = filesplit[0]\n",
    "rep = filesplit[1]\n",
    "\n",
    "print ('Exp type: ' + exp_type)\n",
    "print ('Rep: ' + rep)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Snf2_rep02_MID21_allLanes_tophat2.0.5.bam.gbgout\n",
      "Snf2_rep03_MID28_allLanes_tophat2.0.5.bam.gbgout\n",
      "Snf2_rep04_MID11_allLanes_tophat2.0.5.bam.gbgout\n",
      "Snf2_rep05_MID82_allLanes_tophat2.0.5.bam.gbgout\n",
      "Snf2_rep06_MID80_allLanes_tophat2.0.5.bam.gbgout\n",
      "Snf2_rep07_MID26_allLanes_tophat2.0.5.bam.gbgout\n",
      "Snf2_rep08_MID78_allLanes_tophat2.0.5.bam.gbgout\n",
      "Snf2_rep09_MID04_allLanes_tophat2.0.5.bam.gbgout\n",
      "Snf2_rep10_MID38_allLanes_tophat2.0.5.bam.gbgout\n",
      "Snf2_rep11_MID01_allLanes_tophat2.0.5.bam.gbgout\n",
      "Snf2_rep12_MID02_allLanes_tophat2.0.5.bam.gbgout\n",
      "Snf2_rep13_MID59_allLanes_tophat2.0.5.bam.gbgout\n",
      "Snf2_rep14_MID33_allLanes_tophat2.0.5.bam.gbgout\n",
      "Snf2_rep15_MID75_allLanes_tophat2.0.5.bam.gbgout\n",
      "Snf2_rep16_MID72_allLanes_tophat2.0.5.bam.gbgout\n",
      "Snf2_rep17_MID89_allLanes_tophat2.0.5.bam.gbgout\n",
      "Snf2_rep18_MID40_allLanes_tophat2.0.5.bam.gbgout\n",
      "Snf2_rep19_MID84_allLanes_tophat2.0.5.bam.gbgout\n",
      "Snf2_rep20_MID15_allLanes_tophat2.0.5.bam.gbgout\n",
      "Snf2_rep21_MID31_allLanes_tophat2.0.5.bam.gbgout\n",
      "Snf2_rep22_MID53_allLanes_tophat2.0.5.bam.gbgout\n",
      "Snf2_rep23_MID25_allLanes_tophat2.0.5.bam.gbgout\n",
      "Snf2_rep24_MID42_allLanes_tophat2.0.5.bam.gbgout\n",
      "Snf2_rep25_MID81_allLanes_tophat2.0.5.bam.gbgout\n",
      "Snf2_rep26_MID71_allLanes_tophat2.0.5.bam.gbgout\n",
      "Snf2_rep27_MID34_allLanes_tophat2.0.5.bam.gbgout\n",
      "Snf2_rep28_MID74_allLanes_tophat2.0.5.bam.gbgout\n",
      "Snf2_rep29_MID57_allLanes_tophat2.0.5.bam.gbgout\n",
      "Snf2_rep30_MID16_allLanes_tophat2.0.5.bam.gbgout\n",
      "Snf2_rep31_MID77_allLanes_tophat2.0.5.bam.gbgout\n",
      "Snf2_rep32_MID70_allLanes_tophat2.0.5.bam.gbgout\n",
      "Snf2_rep33_MID29_allLanes_tophat2.0.5.bam.gbgout\n",
      "Snf2_rep34_MID14_allLanes_tophat2.0.5.bam.gbgout\n",
      "Snf2_rep35_MID60_allLanes_tophat2.0.5.bam.gbgout\n",
      "Snf2_rep36_MID23_allLanes_tophat2.0.5.bam.gbgout\n",
      "Snf2_rep37_MID46_allLanes_tophat2.0.5.bam.gbgout\n",
      "Snf2_rep38_MID06_allLanes_tophat2.0.5.bam.gbgout\n",
      "Snf2_rep39_MID19_allLanes_tophat2.0.5.bam.gbgout\n",
      "Snf2_rep40_MID12_allLanes_tophat2.0.5.bam.gbgout\n",
      "Snf2_rep41_MID79_allLanes_tophat2.0.5.bam.gbgout\n",
      "Snf2_rep42_MID47_allLanes_tophat2.0.5.bam.gbgout\n",
      "Snf2_rep43_MID73_allLanes_tophat2.0.5.bam.gbgout\n",
      "Snf2_rep44_MID65_allLanes_tophat2.0.5.bam.gbgout\n",
      "Snf2_rep45_MID86_allLanes_tophat2.0.5.bam.gbgout\n",
      "Snf2_rep46_MID35_allLanes_tophat2.0.5.bam.gbgout\n",
      "Snf2_rep47_MID56_allLanes_tophat2.0.5.bam.gbgout\n",
      "Snf2_rep48_MID55_allLanes_tophat2.0.5.bam.gbgout\n",
      "WT_rep01_MID39_allLanes_tophat2.0.5.bam.gbgout\n",
      "WT_rep02_MID50_allLanes_tophat2.0.5.bam.gbgout\n",
      "WT_rep03_MID51_allLanes_tophat2.0.5.bam.gbgout\n",
      "WT_rep04_MID58_allLanes_tophat2.0.5.bam.gbgout\n",
      "WT_rep05_MID93_allLanes_tophat2.0.5.bam.gbgout\n",
      "WT_rep06_MID30_allLanes_tophat2.0.5.bam.gbgout\n",
      "WT_rep07_MID24_allLanes_tophat2.0.5.bam.gbgout\n",
      "WT_rep08_MID20_allLanes_tophat2.0.5.bam.gbgout\n",
      "WT_rep09_MID91_allLanes_tophat2.0.5.bam.gbgout\n",
      "WT_rep10_MID37_allLanes_tophat2.0.5.bam.gbgout\n",
      "WT_rep11_MID54_allLanes_tophat2.0.5.bam.gbgout\n",
      "WT_rep12_MID03_allLanes_tophat2.0.5.bam.gbgout\n",
      "WT_rep13_MID45_allLanes_tophat2.0.5.bam.gbgout\n",
      "WT_rep14_MID52_allLanes_tophat2.0.5.bam.gbgout\n",
      "WT_rep15_MID09_allLanes_tophat2.0.5.bam.gbgout\n",
      "WT_rep16_MID92_allLanes_tophat2.0.5.bam.gbgout\n",
      "WT_rep17_MID66_allLanes_tophat2.0.5.bam.gbgout\n",
      "WT_rep18_MID48_allLanes_tophat2.0.5.bam.gbgout\n",
      "WT_rep19_MID83_allLanes_tophat2.0.5.bam.gbgout\n",
      "WT_rep20_MID90_allLanes_tophat2.0.5.bam.gbgout\n",
      "WT_rep21_MID62_allLanes_tophat2.0.5.bam.gbgout\n",
      "WT_rep22_MID67_allLanes_tophat2.0.5.bam.gbgout\n",
      "WT_rep23_MID36_allLanes_tophat2.0.5.bam.gbgout\n",
      "WT_rep24_MID69_allLanes_tophat2.0.5.bam.gbgout\n",
      "WT_rep25_MID61_allLanes_tophat2.0.5.bam.gbgout\n",
      "WT_rep26_MID08_allLanes_tophat2.0.5.bam.gbgout\n",
      "WT_rep27_MID41_allLanes_tophat2.0.5.bam.gbgout\n",
      "WT_rep28_MID64_allLanes_tophat2.0.5.bam.gbgout\n",
      "WT_rep29_MID68_allLanes_tophat2.0.5.bam.gbgout\n",
      "WT_rep30_MID27_allLanes_tophat2.0.5.bam.gbgout\n",
      "WT_rep31_MID43_allLanes_tophat2.0.5.bam.gbgout\n",
      "WT_rep32_MID95_allLanes_tophat2.0.5.bam.gbgout\n",
      "WT_rep33_MID88_allLanes_tophat2.0.5.bam.gbgout\n",
      "WT_rep34_MID76_allLanes_tophat2.0.5.bam.gbgout\n",
      "WT_rep35_MID32_allLanes_tophat2.0.5.bam.gbgout\n",
      "WT_rep36_MID63_allLanes_tophat2.0.5.bam.gbgout\n",
      "WT_rep37_MID07_allLanes_tophat2.0.5.bam.gbgout\n",
      "WT_rep38_MID87_allLanes_tophat2.0.5.bam.gbgout\n",
      "WT_rep39_MID05_allLanes_tophat2.0.5.bam.gbgout\n",
      "WT_rep40_MID18_allLanes_tophat2.0.5.bam.gbgout\n",
      "WT_rep41_MID13_allLanes_tophat2.0.5.bam.gbgout\n",
      "WT_rep42_MID10_allLanes_tophat2.0.5.bam.gbgout\n",
      "WT_rep43_MID49_allLanes_tophat2.0.5.bam.gbgout\n",
      "WT_rep44_MID22_allLanes_tophat2.0.5.bam.gbgout\n",
      "WT_rep45_MID44_allLanes_tophat2.0.5.bam.gbgout\n",
      "WT_rep46_MID94_allLanes_tophat2.0.5.bam.gbgout\n",
      "WT_rep47_MID17_allLanes_tophat2.0.5.bam.gbgout\n",
      "WT_rep48_MID85_allLanes_tophat2.0.5.bam.gbgout\n"
     ]
    }
   ],
   "source": [
    "# get a list of files\n",
    "\n",
    "files = [f for f in sorted(os.listdir('data/rawdata')) if f.endswith(\".gbgout\")]\n",
    "\n",
    "# start with first file\n",
    "\n",
    "file = files[0]\n",
    "filesplit = file.split('_')\n",
    "exp_type = filesplit[0]\n",
    "rep = filesplit[1]\n",
    "\n",
    "exp_id = exp_type + \"_\" + rep\n",
    "\n",
    "filepath = os.path.join('data/rawdata/', file)\n",
    "\n",
    "# rather than using Gene, Expression we change the columns to Gene, Experiment ID\n",
    "expression_all = pd.read_csv(filepath, sep=\"\\t\", header = None, names=['Gene',exp_id])\n",
    "\n",
    "# now loop over remainder of files and merge to data frame\n",
    "\n",
    "for file in files[1:]:\n",
    "    print (file)\n",
    "    filesplit = file.split('_')\n",
    "    exp_type = filesplit[0]\n",
    "    rep = filesplit[1]\n",
    "    exp_id = exp_type + \"_\" + rep\n",
    "    filepath = os.path.join('data/rawdata/', file)\n",
    "    \n",
    "    expression_all = expression_all.merge(pd.read_csv(filepath, sep=\"\\t\", header = None, names=['Gene',exp_id]), on=\"Gene\", how=\"outer\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>Snf2_rep01</th>\n",
       "      <th>Snf2_rep02</th>\n",
       "      <th>Snf2_rep03</th>\n",
       "      <th>Snf2_rep04</th>\n",
       "      <th>Snf2_rep05</th>\n",
       "      <th>Snf2_rep06</th>\n",
       "      <th>Snf2_rep07</th>\n",
       "      <th>Snf2_rep08</th>\n",
       "      <th>Snf2_rep09</th>\n",
       "      <th>...</th>\n",
       "      <th>WT_rep39</th>\n",
       "      <th>WT_rep40</th>\n",
       "      <th>WT_rep41</th>\n",
       "      <th>WT_rep42</th>\n",
       "      <th>WT_rep43</th>\n",
       "      <th>WT_rep44</th>\n",
       "      <th>WT_rep45</th>\n",
       "      <th>WT_rep46</th>\n",
       "      <th>WT_rep47</th>\n",
       "      <th>WT_rep48</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>15S_rRNA</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>46</td>\n",
       "      <td>3</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>49</td>\n",
       "      <td>9</td>\n",
       "      <td>4</td>\n",
       "      <td>11</td>\n",
       "      <td>12</td>\n",
       "      <td>1</td>\n",
       "      <td>22</td>\n",
       "      <td>12</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>21S_rRNA</td>\n",
       "      <td>31</td>\n",
       "      <td>18</td>\n",
       "      <td>23</td>\n",
       "      <td>44</td>\n",
       "      <td>356</td>\n",
       "      <td>62</td>\n",
       "      <td>35</td>\n",
       "      <td>33</td>\n",
       "      <td>13</td>\n",
       "      <td>...</td>\n",
       "      <td>10</td>\n",
       "      <td>274</td>\n",
       "      <td>49</td>\n",
       "      <td>30</td>\n",
       "      <td>72</td>\n",
       "      <td>58</td>\n",
       "      <td>21</td>\n",
       "      <td>159</td>\n",
       "      <td>107</td>\n",
       "      <td>70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HRA1</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>...</td>\n",
       "      <td>5</td>\n",
       "      <td>3</td>\n",
       "      <td>6</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ICR1</td>\n",
       "      <td>205</td>\n",
       "      <td>196</td>\n",
       "      <td>211</td>\n",
       "      <td>252</td>\n",
       "      <td>127</td>\n",
       "      <td>146</td>\n",
       "      <td>275</td>\n",
       "      <td>160</td>\n",
       "      <td>190</td>\n",
       "      <td>...</td>\n",
       "      <td>85</td>\n",
       "      <td>177</td>\n",
       "      <td>137</td>\n",
       "      <td>118</td>\n",
       "      <td>113</td>\n",
       "      <td>81</td>\n",
       "      <td>142</td>\n",
       "      <td>94</td>\n",
       "      <td>187</td>\n",
       "      <td>106</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>LSR1</td>\n",
       "      <td>210</td>\n",
       "      <td>103</td>\n",
       "      <td>159</td>\n",
       "      <td>260</td>\n",
       "      <td>298</td>\n",
       "      <td>522</td>\n",
       "      <td>303</td>\n",
       "      <td>96</td>\n",
       "      <td>132</td>\n",
       "      <td>...</td>\n",
       "      <td>66</td>\n",
       "      <td>385</td>\n",
       "      <td>232</td>\n",
       "      <td>149</td>\n",
       "      <td>114</td>\n",
       "      <td>81</td>\n",
       "      <td>109</td>\n",
       "      <td>132</td>\n",
       "      <td>243</td>\n",
       "      <td>128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7126</th>\n",
       "      <td>no_feature</td>\n",
       "      <td>888995</td>\n",
       "      <td>641333</td>\n",
       "      <td>544060</td>\n",
       "      <td>955003</td>\n",
       "      <td>616339</td>\n",
       "      <td>764571</td>\n",
       "      <td>1290592</td>\n",
       "      <td>598769</td>\n",
       "      <td>827934</td>\n",
       "      <td>...</td>\n",
       "      <td>403527</td>\n",
       "      <td>1035268</td>\n",
       "      <td>759722</td>\n",
       "      <td>578053</td>\n",
       "      <td>615809</td>\n",
       "      <td>446796</td>\n",
       "      <td>683906</td>\n",
       "      <td>462390</td>\n",
       "      <td>901045</td>\n",
       "      <td>639100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7127</th>\n",
       "      <td>ambiguous</td>\n",
       "      <td>730793</td>\n",
       "      <td>665470</td>\n",
       "      <td>709428</td>\n",
       "      <td>821933</td>\n",
       "      <td>1171511</td>\n",
       "      <td>529946</td>\n",
       "      <td>925353</td>\n",
       "      <td>669715</td>\n",
       "      <td>717502</td>\n",
       "      <td>...</td>\n",
       "      <td>583744</td>\n",
       "      <td>2002465</td>\n",
       "      <td>865662</td>\n",
       "      <td>646805</td>\n",
       "      <td>570375</td>\n",
       "      <td>1011281</td>\n",
       "      <td>899586</td>\n",
       "      <td>596335</td>\n",
       "      <td>1321941</td>\n",
       "      <td>803015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7128</th>\n",
       "      <td>too_low_aQual</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7129</th>\n",
       "      <td>not_aligned</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7130</th>\n",
       "      <td>alignment_not_unique</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7131 rows × 97 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      Gene  Snf2_rep01  Snf2_rep02  Snf2_rep03  Snf2_rep04  \\\n",
       "0                 15S_rRNA           4           2           5           5   \n",
       "1                 21S_rRNA          31          18          23          44   \n",
       "2                     HRA1           5           1           3           1   \n",
       "3                     ICR1         205         196         211         252   \n",
       "4                     LSR1         210         103         159         260   \n",
       "...                    ...         ...         ...         ...         ...   \n",
       "7126            no_feature      888995      641333      544060      955003   \n",
       "7127             ambiguous      730793      665470      709428      821933   \n",
       "7128         too_low_aQual           0           0           0           0   \n",
       "7129           not_aligned           0           0           0           0   \n",
       "7130  alignment_not_unique           0           0           0           0   \n",
       "\n",
       "      Snf2_rep05  Snf2_rep06  Snf2_rep07  Snf2_rep08  Snf2_rep09  ...  \\\n",
       "0             46           3           5           4           2  ...   \n",
       "1            356          62          35          33          13  ...   \n",
       "2              2           1           1           4           4  ...   \n",
       "3            127         146         275         160         190  ...   \n",
       "4            298         522         303          96         132  ...   \n",
       "...          ...         ...         ...         ...         ...  ...   \n",
       "7126      616339      764571     1290592      598769      827934  ...   \n",
       "7127     1171511      529946      925353      669715      717502  ...   \n",
       "7128           0           0           0           0           0  ...   \n",
       "7129           0           0           0           0           0  ...   \n",
       "7130           0           0           0           0           0  ...   \n",
       "\n",
       "      WT_rep39  WT_rep40  WT_rep41  WT_rep42  WT_rep43  WT_rep44  WT_rep45  \\\n",
       "0            0        49         9         4        11        12         1   \n",
       "1           10       274        49        30        72        58        21   \n",
       "2            5         3         6         5         2         2         2   \n",
       "3           85       177       137       118       113        81       142   \n",
       "4           66       385       232       149       114        81       109   \n",
       "...        ...       ...       ...       ...       ...       ...       ...   \n",
       "7126    403527   1035268    759722    578053    615809    446796    683906   \n",
       "7127    583744   2002465    865662    646805    570375   1011281    899586   \n",
       "7128         0         0         0         0         0         0         0   \n",
       "7129         0         0         0         0         0         0         0   \n",
       "7130         0         0         0         0         0         0         0   \n",
       "\n",
       "      WT_rep46  WT_rep47  WT_rep48  \n",
       "0           22        12         4  \n",
       "1          159       107        70  \n",
       "2            5         2         1  \n",
       "3           94       187       106  \n",
       "4          132       243       128  \n",
       "...        ...       ...       ...  \n",
       "7126    462390    901045    639100  \n",
       "7127    596335   1321941    803015  \n",
       "7128         0         0         0  \n",
       "7129         0         0         0  \n",
       "7130         0         0         0  \n",
       "\n",
       "[7131 rows x 97 columns]"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "expression_all"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The files are incredible simple. Two columns, separated by tabs. The first column is the gene name, the second column the normalized counts for that gene in that sample (read the paper if you're interested in how they did the counting and normalization).\n",
    "\n",
    "YBL013W 39 YBL014C 127 YBL015W 732 YBL016W 309 YBL017C 1613 YBL018C 174 YBL019W 117 YBL020W 258 YBL021C 248 YBL022C 1168 YBL023C 331 YBL024W 451 YBL025W 64 YBL026W 206 YBL027W 9723 YBL028C 77 YBL029C-A 157\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "expression_all = expression_all[~expression_all.Gene.isin(rows_to_drop)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>Snf2_rep01</th>\n",
       "      <th>Snf2_rep02</th>\n",
       "      <th>Snf2_rep03</th>\n",
       "      <th>Snf2_rep04</th>\n",
       "      <th>Snf2_rep05</th>\n",
       "      <th>Snf2_rep06</th>\n",
       "      <th>Snf2_rep07</th>\n",
       "      <th>Snf2_rep08</th>\n",
       "      <th>Snf2_rep09</th>\n",
       "      <th>...</th>\n",
       "      <th>WT_rep39</th>\n",
       "      <th>WT_rep40</th>\n",
       "      <th>WT_rep41</th>\n",
       "      <th>WT_rep42</th>\n",
       "      <th>WT_rep43</th>\n",
       "      <th>WT_rep44</th>\n",
       "      <th>WT_rep45</th>\n",
       "      <th>WT_rep46</th>\n",
       "      <th>WT_rep47</th>\n",
       "      <th>WT_rep48</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>15S_rRNA</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>46</td>\n",
       "      <td>3</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>49</td>\n",
       "      <td>9</td>\n",
       "      <td>4</td>\n",
       "      <td>11</td>\n",
       "      <td>12</td>\n",
       "      <td>1</td>\n",
       "      <td>22</td>\n",
       "      <td>12</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>21S_rRNA</td>\n",
       "      <td>31</td>\n",
       "      <td>18</td>\n",
       "      <td>23</td>\n",
       "      <td>44</td>\n",
       "      <td>356</td>\n",
       "      <td>62</td>\n",
       "      <td>35</td>\n",
       "      <td>33</td>\n",
       "      <td>13</td>\n",
       "      <td>...</td>\n",
       "      <td>10</td>\n",
       "      <td>274</td>\n",
       "      <td>49</td>\n",
       "      <td>30</td>\n",
       "      <td>72</td>\n",
       "      <td>58</td>\n",
       "      <td>21</td>\n",
       "      <td>159</td>\n",
       "      <td>107</td>\n",
       "      <td>70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HRA1</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>...</td>\n",
       "      <td>5</td>\n",
       "      <td>3</td>\n",
       "      <td>6</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ICR1</td>\n",
       "      <td>205</td>\n",
       "      <td>196</td>\n",
       "      <td>211</td>\n",
       "      <td>252</td>\n",
       "      <td>127</td>\n",
       "      <td>146</td>\n",
       "      <td>275</td>\n",
       "      <td>160</td>\n",
       "      <td>190</td>\n",
       "      <td>...</td>\n",
       "      <td>85</td>\n",
       "      <td>177</td>\n",
       "      <td>137</td>\n",
       "      <td>118</td>\n",
       "      <td>113</td>\n",
       "      <td>81</td>\n",
       "      <td>142</td>\n",
       "      <td>94</td>\n",
       "      <td>187</td>\n",
       "      <td>106</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>LSR1</td>\n",
       "      <td>210</td>\n",
       "      <td>103</td>\n",
       "      <td>159</td>\n",
       "      <td>260</td>\n",
       "      <td>298</td>\n",
       "      <td>522</td>\n",
       "      <td>303</td>\n",
       "      <td>96</td>\n",
       "      <td>132</td>\n",
       "      <td>...</td>\n",
       "      <td>66</td>\n",
       "      <td>385</td>\n",
       "      <td>232</td>\n",
       "      <td>149</td>\n",
       "      <td>114</td>\n",
       "      <td>81</td>\n",
       "      <td>109</td>\n",
       "      <td>132</td>\n",
       "      <td>243</td>\n",
       "      <td>128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7121</th>\n",
       "      <td>tY(GUA)J2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7122</th>\n",
       "      <td>tY(GUA)M1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7123</th>\n",
       "      <td>tY(GUA)M2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7124</th>\n",
       "      <td>tY(GUA)O</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7125</th>\n",
       "      <td>tY(GUA)Q</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7126 rows × 97 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           Gene  Snf2_rep01  Snf2_rep02  Snf2_rep03  Snf2_rep04  Snf2_rep05  \\\n",
       "0      15S_rRNA           4           2           5           5          46   \n",
       "1      21S_rRNA          31          18          23          44         356   \n",
       "2          HRA1           5           1           3           1           2   \n",
       "3          ICR1         205         196         211         252         127   \n",
       "4          LSR1         210         103         159         260         298   \n",
       "...         ...         ...         ...         ...         ...         ...   \n",
       "7121  tY(GUA)J2           1           0           0           0           0   \n",
       "7122  tY(GUA)M1           0           0           1           0           2   \n",
       "7123  tY(GUA)M2           0           0           0           0           0   \n",
       "7124   tY(GUA)O           0           1           0           0           0   \n",
       "7125   tY(GUA)Q           0           0           0           0           0   \n",
       "\n",
       "      Snf2_rep06  Snf2_rep07  Snf2_rep08  Snf2_rep09  ...  WT_rep39  WT_rep40  \\\n",
       "0              3           5           4           2  ...         0        49   \n",
       "1             62          35          33          13  ...        10       274   \n",
       "2              1           1           4           4  ...         5         3   \n",
       "3            146         275         160         190  ...        85       177   \n",
       "4            522         303          96         132  ...        66       385   \n",
       "...          ...         ...         ...         ...  ...       ...       ...   \n",
       "7121           0           0           1           0  ...         0         0   \n",
       "7122           0           0           0           0  ...         0         1   \n",
       "7123           0           0           0           0  ...         0         0   \n",
       "7124           0           0           0           0  ...         0         0   \n",
       "7125           0           0           0           0  ...         0         0   \n",
       "\n",
       "      WT_rep41  WT_rep42  WT_rep43  WT_rep44  WT_rep45  WT_rep46  WT_rep47  \\\n",
       "0            9         4        11        12         1        22        12   \n",
       "1           49        30        72        58        21       159       107   \n",
       "2            6         5         2         2         2         5         2   \n",
       "3          137       118       113        81       142        94       187   \n",
       "4          232       149       114        81       109       132       243   \n",
       "...        ...       ...       ...       ...       ...       ...       ...   \n",
       "7121         0         0         0         0         0         0         0   \n",
       "7122         1         0         1         0         0         0         1   \n",
       "7123         0         0         0         0         0         0         0   \n",
       "7124         1         0         0         0         0         0         0   \n",
       "7125         0         0         0         0         0         0         0   \n",
       "\n",
       "      WT_rep48  \n",
       "0            4  \n",
       "1           70  \n",
       "2            1  \n",
       "3          106  \n",
       "4          128  \n",
       "...        ...  \n",
       "7121         0  \n",
       "7122         0  \n",
       "7123         0  \n",
       "7124         1  \n",
       "7125         0  \n",
       "\n",
       "[7126 rows x 97 columns]"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "expression_all"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Save\n",
    "\n",
    "expression_all.to_csv(\"Barton_combined.txt\", sep='\\t')"
   ]
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
