{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "from functools import reduce\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "from scipy.stats.stats import pearsonr\n",
    "from scipy import stats\n",
    "from scipy.stats import ks_2samp\n",
    "from scipy.stats import entropy"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Datafiles\n",
    "There are 96 separate datafiles, each representing the processed output of an RNAseq experiment looking at mRNA levels in wild-type (48 samples) or SNF2 deletion (48 samples) yeast cells grown under standard condition."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Reading individual files with Pandas\n",
    "\n",
    "The individual files are in the directory \"data/rawdata/\" and are tab-delimeted text files. These are basically raw spreadsheets stored as text, with individual lines separated by a **'\\n'** which indicated 'new line', and columns within each line separated by a **'\\t'** or 'tab' character. (We will also deal with files where the column is separated by a comma."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Like any good program for working with spreadsheet data, Pandas knows how to read this kind of file if we tell it what to look for. To do this we use its [read_csv](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_csv.html) function, specifying first the filepath, and then a few details about this file.  \n",
    "\n",
    "**sep=\"\\t\"** tells Pandas that the data separator is a tab character (as opposed to other common separators like comma or space)  \n",
    "**header=None** tells Pandas that the first line of the file is not a header (this is what it typically expects)  \n",
    "**names = ..** because there is no header, we tell Pandas what to call the columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>Expression</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>15S_rRNA</td>\n",
       "      <td>29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>21S_rRNA</td>\n",
       "      <td>221</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HRA1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ICR1</td>\n",
       "      <td>70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>LSR1</td>\n",
       "      <td>314</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7126</th>\n",
       "      <td>no_feature</td>\n",
       "      <td>414503</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7127</th>\n",
       "      <td>ambiguous</td>\n",
       "      <td>1387186</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7128</th>\n",
       "      <td>too_low_aQual</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7129</th>\n",
       "      <td>not_aligned</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7130</th>\n",
       "      <td>alignment_not_unique</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7131 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      Gene  Expression\n",
       "0                 15S_rRNA          29\n",
       "1                 21S_rRNA         221\n",
       "2                     HRA1           0\n",
       "3                     ICR1          70\n",
       "4                     LSR1         314\n",
       "...                    ...         ...\n",
       "7126            no_feature      414503\n",
       "7127             ambiguous     1387186\n",
       "7128         too_low_aQual           0\n",
       "7129           not_aligned           0\n",
       "7130  alignment_not_unique           0\n",
       "\n",
       "[7131 rows x 2 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.read_csv(\"data/rawdata/WT_rep16_MID92_allLanes_tophat2.0.5.bam.gbgout\",sep=\"\\t\",header=None,names=[\"Gene\",\"Expression\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If we just run that command, Pandas loads in the file and displays it. But we want to store the data, so we assign the data to a variable. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "ge = pd.read_csv(\"data/rawdata/WT_rep16_MID92_allLanes_tophat2.0.5.bam.gbgout\",sep=\"\\t\",header=None,names=[\"Gene\",\"Expression\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we can look at the file in different ways. The **head(n)** command shows us the first n rows."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>Expression</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>15S_rRNA</td>\n",
       "      <td>29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>21S_rRNA</td>\n",
       "      <td>221</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HRA1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ICR1</td>\n",
       "      <td>70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>LSR1</td>\n",
       "      <td>314</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>NME1</td>\n",
       "      <td>33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>PWR1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Q0010</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Q0017</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Q0032</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Gene  Expression\n",
       "0  15S_rRNA          29\n",
       "1  21S_rRNA         221\n",
       "2      HRA1           0\n",
       "3      ICR1          70\n",
       "4      LSR1         314\n",
       "5      NME1          33\n",
       "6      PWR1           0\n",
       "7     Q0010           0\n",
       "8     Q0017           0\n",
       "9     Q0032           0"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ge.head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Similarly the **tail(n)** command shows us the last n rows."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>Expression</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>7121</th>\n",
       "      <td>tY(GUA)J2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7122</th>\n",
       "      <td>tY(GUA)M1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7123</th>\n",
       "      <td>tY(GUA)M2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7124</th>\n",
       "      <td>tY(GUA)O</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7125</th>\n",
       "      <td>tY(GUA)Q</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7126</th>\n",
       "      <td>no_feature</td>\n",
       "      <td>414503</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7127</th>\n",
       "      <td>ambiguous</td>\n",
       "      <td>1387186</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7128</th>\n",
       "      <td>too_low_aQual</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7129</th>\n",
       "      <td>not_aligned</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7130</th>\n",
       "      <td>alignment_not_unique</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      Gene  Expression\n",
       "7121             tY(GUA)J2           0\n",
       "7122             tY(GUA)M1           0\n",
       "7123             tY(GUA)M2           0\n",
       "7124              tY(GUA)O           0\n",
       "7125              tY(GUA)Q           0\n",
       "7126            no_feature      414503\n",
       "7127             ambiguous     1387186\n",
       "7128         too_low_aQual           0\n",
       "7129           not_aligned           0\n",
       "7130  alignment_not_unique           0"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ge.tail(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "You can already see one (typical) problem, which is that the file is not just data. It also how some summary statistics that we don't care about. So we have to filter the rows in some manner. Here we use a bit of data about gene nomenclature in yeast. \n",
    "\n",
    "Most of the genes have this naming format:\n",
    "\n",
    "Systematic names for nuclear-encoded ORFs begin with the letter 'Y' (for 'Yeast'); the second letter denotes the chromosome number ('A' is chr I, 'B' is chr II, etc.); the third letter is either 'L' or 'R' for left or right chromosome arm; next is a three digit number indicating the order of the ORFs on that arm of a chromosome starting from the centromere, irrespective of strand; finally, there is an additional letter indicating the strand, either 'W' for Watson (the strand with 5' end at the left telomere) or 'C' for Crick (the complement strand, 5' end is at the right telomere).\n",
    "\n",
    "You can read about the yeast gene naming system here - http://seq.yeastgenome.org/help/community/nomenclature-conventions."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can look for rows that match this style by using the built in filtering system in Pandas. This specific command asks Pandas to return all rows where the Gene column starts with the character \"Y\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>Expression</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>YAL001C</td>\n",
       "      <td>232</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>YAL002W</td>\n",
       "      <td>245</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>YAL003W</td>\n",
       "      <td>8073</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>YAL004W</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>YAL005C</td>\n",
       "      <td>7471</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6741</th>\n",
       "      <td>YPR201W</td>\n",
       "      <td>45</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6742</th>\n",
       "      <td>YPR202W</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6743</th>\n",
       "      <td>YPR203W</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6744</th>\n",
       "      <td>YPR204C-A</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6745</th>\n",
       "      <td>YPR204W</td>\n",
       "      <td>633</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6685 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           Gene  Expression\n",
       "61      YAL001C         232\n",
       "62      YAL002W         245\n",
       "63      YAL003W        8073\n",
       "64      YAL004W           0\n",
       "65      YAL005C        7471\n",
       "...         ...         ...\n",
       "6741    YPR201W          45\n",
       "6742    YPR202W          22\n",
       "6743    YPR203W          18\n",
       "6744  YPR204C-A           0\n",
       "6745    YPR204W         633\n",
       "\n",
       "[6685 rows x 2 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ge[ge.Gene.str.startswith(\"Y\")]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "That accounts for 6685 of the 7130 rows. But what about the rest? We can start by reversing that filter to ask for rows that do not start with \"Y\". That work is done here with the **~** symbol."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>Expression</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>15S_rRNA</td>\n",
       "      <td>29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>21S_rRNA</td>\n",
       "      <td>221</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HRA1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ICR1</td>\n",
       "      <td>70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>LSR1</td>\n",
       "      <td>314</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7126</th>\n",
       "      <td>no_feature</td>\n",
       "      <td>414503</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7127</th>\n",
       "      <td>ambiguous</td>\n",
       "      <td>1387186</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7128</th>\n",
       "      <td>too_low_aQual</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7129</th>\n",
       "      <td>not_aligned</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7130</th>\n",
       "      <td>alignment_not_unique</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>446 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      Gene  Expression\n",
       "0                 15S_rRNA          29\n",
       "1                 21S_rRNA         221\n",
       "2                     HRA1           0\n",
       "3                     ICR1          70\n",
       "4                     LSR1         314\n",
       "...                    ...         ...\n",
       "7126            no_feature      414503\n",
       "7127             ambiguous     1387186\n",
       "7128         too_low_aQual           0\n",
       "7129           not_aligned           0\n",
       "7130  alignment_not_unique           0\n",
       "\n",
       "[446 rows x 2 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ge[~ge.Gene.str.startswith(\"Y\")]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This is going to look a bit messy, but we can ask Pandas to return all of the Gene names for these remaining rows, just so we can get a sense of what we're dealing with."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['15S_rRNA', '21S_rRNA', 'HRA1', 'ICR1', 'LSR1', 'NME1', 'PWR1',\n",
       "       'Q0010', 'Q0017', 'Q0032', 'Q0045', 'Q0050', 'Q0055', 'Q0060',\n",
       "       'Q0065', 'Q0070', 'Q0075', 'Q0080', 'Q0085', 'Q0092', 'Q0105',\n",
       "       'Q0110', 'Q0115', 'Q0120', 'Q0130', 'Q0140', 'Q0142', 'Q0143',\n",
       "       'Q0144', 'Q0160', 'Q0182', 'Q0250', 'Q0255', 'Q0275', 'Q0297',\n",
       "       'RDN18-1', 'RDN18-2', 'RDN25-1', 'RDN25-2', 'RDN37-1', 'RDN37-2',\n",
       "       'RDN5-1', 'RDN5-2', 'RDN5-3', 'RDN5-4', 'RDN5-5', 'RDN5-6',\n",
       "       'RDN58-1', 'RDN58-2', 'RNA170', 'RPM1', 'RPR1', 'RUF20', 'RUF21',\n",
       "       'RUF22', 'RUF23', 'RUF5-1', 'RUF5-2', 'SCR1', 'SRG1', 'TLC1',\n",
       "       'snR10', 'snR11', 'snR128', 'snR13', 'snR14', 'snR161', 'snR17a',\n",
       "       'snR17b', 'snR18', 'snR189', 'snR19', 'snR190', 'snR191', 'snR24',\n",
       "       'snR3', 'snR30', 'snR31', 'snR32', 'snR33', 'snR34', 'snR35',\n",
       "       'snR36', 'snR37', 'snR38', 'snR39', 'snR39B', 'snR4', 'snR40',\n",
       "       'snR41', 'snR42', 'snR43', 'snR44', 'snR45', 'snR46', 'snR47',\n",
       "       'snR48', 'snR49', 'snR5', 'snR50', 'snR51', 'snR52', 'snR53',\n",
       "       'snR54', 'snR55', 'snR56', 'snR57', 'snR58', 'snR59', 'snR6',\n",
       "       'snR60', 'snR61', 'snR62', 'snR63', 'snR64', 'snR65', 'snR66',\n",
       "       'snR67', 'snR68', 'snR69', 'snR7-L', 'snR7-S', 'snR70', 'snR71',\n",
       "       'snR72', 'snR73', 'snR74', 'snR75', 'snR76', 'snR77', 'snR78',\n",
       "       'snR79', 'snR8', 'snR80', 'snR81', 'snR82', 'snR83', 'snR84',\n",
       "       'snR85', 'snR86', 'snR87', 'snR9', 'tA(AGC)D', 'tA(AGC)F',\n",
       "       'tA(AGC)G', 'tA(AGC)H', 'tA(AGC)J', 'tA(AGC)K1', 'tA(AGC)K2',\n",
       "       'tA(AGC)L', 'tA(AGC)M1', 'tA(AGC)M2', 'tA(AGC)P', 'tA(UGC)A',\n",
       "       'tA(UGC)E', 'tA(UGC)G', 'tA(UGC)L', 'tA(UGC)O', 'tA(UGC)Q',\n",
       "       'tC(GCA)B', 'tC(GCA)G', 'tC(GCA)P1', 'tC(GCA)P2', 'tC(GCA)Q',\n",
       "       'tD(GUC)B', 'tD(GUC)D', 'tD(GUC)G1', 'tD(GUC)G2', 'tD(GUC)I1',\n",
       "       'tD(GUC)I2', 'tD(GUC)J1', 'tD(GUC)J2', 'tD(GUC)J3', 'tD(GUC)J4',\n",
       "       'tD(GUC)K', 'tD(GUC)L1', 'tD(GUC)L2', 'tD(GUC)M', 'tD(GUC)N',\n",
       "       'tD(GUC)O', 'tD(GUC)Q', 'tE(CUC)D', 'tE(CUC)I', 'tE(UUC)B',\n",
       "       'tE(UUC)C', 'tE(UUC)E1', 'tE(UUC)E2', 'tE(UUC)E3', 'tE(UUC)G1',\n",
       "       'tE(UUC)G2', 'tE(UUC)G3', 'tE(UUC)I', 'tE(UUC)J', 'tE(UUC)K',\n",
       "       'tE(UUC)L', 'tE(UUC)M', 'tE(UUC)P', 'tE(UUC)Q', 'tF(GAA)B',\n",
       "       'tF(GAA)D', 'tF(GAA)F', 'tF(GAA)G', 'tF(GAA)H1', 'tF(GAA)H2',\n",
       "       'tF(GAA)M', 'tF(GAA)N', 'tF(GAA)P1', 'tF(GAA)P2', 'tF(GAA)Q',\n",
       "       'tG(CCC)D', 'tG(CCC)O', 'tG(GCC)B', 'tG(GCC)C', 'tG(GCC)D1',\n",
       "       'tG(GCC)D2', 'tG(GCC)E', 'tG(GCC)F1', 'tG(GCC)F2', 'tG(GCC)G1',\n",
       "       'tG(GCC)G2', 'tG(GCC)J1', 'tG(GCC)J2', 'tG(GCC)M', 'tG(GCC)O1',\n",
       "       'tG(GCC)O2', 'tG(GCC)P1', 'tG(GCC)P2', 'tG(UCC)G', 'tG(UCC)N',\n",
       "       'tG(UCC)O', 'tG(UCC)Q', 'tH(GUG)E1', 'tH(GUG)E2', 'tH(GUG)G1',\n",
       "       'tH(GUG)G2', 'tH(GUG)H', 'tH(GUG)K', 'tH(GUG)M', 'tH(GUG)Q',\n",
       "       'tI(AAU)B', 'tI(AAU)D', 'tI(AAU)E1', 'tI(AAU)E2', 'tI(AAU)G',\n",
       "       'tI(AAU)I1', 'tI(AAU)I2', 'tI(AAU)L1', 'tI(AAU)L2', 'tI(AAU)N1',\n",
       "       'tI(AAU)N2', 'tI(AAU)P1', 'tI(AAU)P2', 'tI(GAU)Q', 'tI(UAU)D',\n",
       "       'tI(UAU)L', 'tK(CUU)C', 'tK(CUU)D1', 'tK(CUU)D2', 'tK(CUU)E1',\n",
       "       'tK(CUU)E2', 'tK(CUU)F', 'tK(CUU)G1', 'tK(CUU)G2', 'tK(CUU)G3',\n",
       "       'tK(CUU)I', 'tK(CUU)J', 'tK(CUU)K', 'tK(CUU)M', 'tK(CUU)P',\n",
       "       'tK(UUU)D', 'tK(UUU)G1', 'tK(UUU)G2', 'tK(UUU)K', 'tK(UUU)L',\n",
       "       'tK(UUU)O', 'tK(UUU)P', 'tK(UUU)Q', 'tL(CAA)A', 'tL(CAA)C',\n",
       "       'tL(CAA)D', 'tL(CAA)G1', 'tL(CAA)G2', 'tL(CAA)G3', 'tL(CAA)K',\n",
       "       'tL(CAA)L', 'tL(CAA)M', 'tL(CAA)N', 'tL(GAG)G', 'tL(UAA)B1',\n",
       "       'tL(UAA)B2', 'tL(UAA)D', 'tL(UAA)J', 'tL(UAA)K', 'tL(UAA)L',\n",
       "       'tL(UAA)N', 'tL(UAA)Q', 'tL(UAG)J', 'tL(UAG)L1', 'tL(UAG)L2',\n",
       "       'tM(CAU)C', 'tM(CAU)D', 'tM(CAU)E', 'tM(CAU)J1', 'tM(CAU)J2',\n",
       "       'tM(CAU)J3', 'tM(CAU)M', 'tM(CAU)O1', 'tM(CAU)O2', 'tM(CAU)P',\n",
       "       'tM(CAU)Q1', 'tM(CAU)Q2', 'tN(GUU)C', 'tN(GUU)F', 'tN(GUU)G',\n",
       "       'tN(GUU)K', 'tN(GUU)L', 'tN(GUU)N1', 'tN(GUU)N2', 'tN(GUU)O1',\n",
       "       'tN(GUU)O2', 'tN(GUU)P', 'tN(GUU)Q', 'tP(AGG)C', 'tP(AGG)N',\n",
       "       'tP(UGG)A', 'tP(UGG)F', 'tP(UGG)H', 'tP(UGG)L', 'tP(UGG)M',\n",
       "       'tP(UGG)N1', 'tP(UGG)N2', 'tP(UGG)O1', 'tP(UGG)O2', 'tP(UGG)O3',\n",
       "       'tP(UGG)Q', 'tQ(CUG)M', 'tQ(UUG)B', 'tQ(UUG)C', 'tQ(UUG)D1',\n",
       "       'tQ(UUG)D2', 'tQ(UUG)D3', 'tQ(UUG)E1', 'tQ(UUG)E2', 'tQ(UUG)H',\n",
       "       'tQ(UUG)L', 'tQ(UUG)Q', 'tR(ACG)D', 'tR(ACG)E', 'tR(ACG)J',\n",
       "       'tR(ACG)K', 'tR(ACG)L', 'tR(ACG)O', 'tR(ACG)Q2', 'tR(CCG)L',\n",
       "       'tR(CCU)J', 'tR(UCU)B', 'tR(UCU)D', 'tR(UCU)E', 'tR(UCU)G1',\n",
       "       'tR(UCU)G2', 'tR(UCU)G3', 'tR(UCU)J1', 'tR(UCU)J2', 'tR(UCU)K',\n",
       "       'tR(UCU)M1', 'tR(UCU)M2', 'tR(UCU)Q1', 'tS(AGA)A', 'tS(AGA)B',\n",
       "       'tS(AGA)D1', 'tS(AGA)D2', 'tS(AGA)D3', 'tS(AGA)E', 'tS(AGA)G',\n",
       "       'tS(AGA)H', 'tS(AGA)J', 'tS(AGA)L', 'tS(AGA)M', 'tS(CGA)C',\n",
       "       'tS(GCU)F', 'tS(GCU)L', 'tS(GCU)O', 'tS(GCU)Q1', 'tS(UGA)E',\n",
       "       'tS(UGA)I', 'tS(UGA)P', 'tS(UGA)Q2', 'tT(AGU)B', 'tT(AGU)C',\n",
       "       'tT(AGU)D', 'tT(AGU)H', 'tT(AGU)I1', 'tT(AGU)I2', 'tT(AGU)J',\n",
       "       'tT(AGU)N1', 'tT(AGU)N2', 'tT(AGU)O1', 'tT(AGU)O2', 'tT(CGU)K',\n",
       "       'tT(UGU)G1', 'tT(UGU)G2', 'tT(UGU)H', 'tT(UGU)P', 'tT(UGU)Q1',\n",
       "       'tT(XXX)Q2', 'tV(AAC)E1', 'tV(AAC)E2', 'tV(AAC)G1', 'tV(AAC)G2',\n",
       "       'tV(AAC)G3', 'tV(AAC)H', 'tV(AAC)J', 'tV(AAC)K1', 'tV(AAC)K2',\n",
       "       'tV(AAC)L', 'tV(AAC)M1', 'tV(AAC)M2', 'tV(AAC)M3', 'tV(AAC)O',\n",
       "       'tV(CAC)D', 'tV(CAC)H', 'tV(UAC)B', 'tV(UAC)D', 'tV(UAC)Q',\n",
       "       'tW(CCA)G1', 'tW(CCA)G2', 'tW(CCA)J', 'tW(CCA)K', 'tW(CCA)M',\n",
       "       'tW(CCA)P', 'tW(UCA)Q', 'tX(XXX)D', 'tY(GUA)D', 'tY(GUA)F1',\n",
       "       'tY(GUA)F2', 'tY(GUA)J1', 'tY(GUA)J2', 'tY(GUA)M1', 'tY(GUA)M2',\n",
       "       'tY(GUA)O', 'tY(GUA)Q', 'no_feature', 'ambiguous', 'too_low_aQual',\n",
       "       'not_aligned', 'alignment_not_unique'], dtype=object)"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ge[~ge.Gene.str.startswith(\"Y\")].Gene.values"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "There are several types of things in here.  \n",
    "  \n",
    "**ribosomal RNAs** '15S_rRNA', '21S_rRNA'  \n",
    "**non-coding RNAs** 'HRA1', 'ICR1', 'LSR1', 'NME1', 'PWR1', etc..    \n",
    "**dubious open-reading frames** 'Q0010', 'Q0017', 'Q0032', 'Q0045', etc...  \n",
    "**snoRNAs** snR10', 'snR11', 'snR128', 'snR13', 'snR14', 'snR161', 'snR17a', etc...  \n",
    "**tRNAs** 'tA(AGC)D', 'tA(AGC)F','tA(AGC)G', 'tA(AGC)H', 'tA(AGC)J', 'tA(AGC)K1', etc...  \n",
    "**summary rows - not genes** 'no_feature', 'ambiguous', 'too_low_aQual','not_aligned', 'alignment_not_unique'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For now, we're just going to not use these, but you are free to come back later and incorporate them into our analyses."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "ge = ge[ge.Gene.str.startswith(\"Y\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>Expression</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>YAL001C</td>\n",
       "      <td>232</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>YAL002W</td>\n",
       "      <td>245</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>YAL003W</td>\n",
       "      <td>8073</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>YAL004W</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>YAL005C</td>\n",
       "      <td>7471</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>YAL007C</td>\n",
       "      <td>550</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>YAL008W</td>\n",
       "      <td>228</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>YAL009W</td>\n",
       "      <td>155</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>YAL010C</td>\n",
       "      <td>176</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>YAL011W</td>\n",
       "      <td>203</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Gene  Expression\n",
       "61  YAL001C         232\n",
       "62  YAL002W         245\n",
       "63  YAL003W        8073\n",
       "64  YAL004W           0\n",
       "65  YAL005C        7471\n",
       "66  YAL007C         550\n",
       "67  YAL008W         228\n",
       "68  YAL009W         155\n",
       "69  YAL010C         176\n",
       "70  YAL011W         203"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ge.head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "So, we've got 96 files like this. We want to com"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "ge1 = pd.read_csv(\"data/rawdata/WT_rep16_MID92_allLanes_tophat2.0.5.bam.gbgout\",sep=\"\\t\",header=None,names=[\"Gene\",\"Expression\"])\n",
    "ge2 = pd.read_csv(\"data/rawdata/WT_rep16_MID92_allLanes_tophat2.0.5.bam.gbgout\",sep=\"\\t\",header=None,names=[\"Gene\",\"Expression\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The files are incredible simple. Two columns, separated by tabs. The first column is the gene name, the second column the normalized counts for that gene in that sample (read the paper if you're interested in how they did the counting and normalization).\n",
    "\n",
    "YBL013W 39 YBL014C 127 YBL015W 732 YBL016W 309 YBL017C 1613 YBL018C 174 YBL019W 117 YBL020W 258 YBL021C 248 YBL022C 1168 YBL023C 331 YBL024W 451 YBL025W 64 YBL026W 206 YBL027W 9723 YBL028C 77 YBL029C-A 157\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>15S_rRNA</td>\n",
       "      <td>29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>21S_rRNA</td>\n",
       "      <td>221</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HRA1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ICR1</td>\n",
       "      <td>70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>LSR1</td>\n",
       "      <td>314</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>NME1</td>\n",
       "      <td>33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>PWR1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Q0010</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Q0017</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Q0032</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          0    1\n",
       "0  15S_rRNA   29\n",
       "1  21S_rRNA  221\n",
       "2      HRA1    0\n",
       "3      ICR1   70\n",
       "4      LSR1  314\n",
       "5      NME1   33\n",
       "6      PWR1    0\n",
       "7     Q0010    0\n",
       "8     Q0017    0\n",
       "9     Q0032    0"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv(\"rawdata/WT_rep16_MID92_allLanes_tophat2.0.5.bam.gbgout\",sep=\"\\t\",header=None,names=[\"Gene\",\"Expression\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>Expression</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>7130</th>\n",
       "      <td>alignment_not_unique</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>444</th>\n",
       "      <td>YBR121C-A</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6717</th>\n",
       "      <td>YPR177C</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6709</th>\n",
       "      <td>YPR170W-A</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3166</th>\n",
       "      <td>YIL082W</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>440</th>\n",
       "      <td>YBR118W</td>\n",
       "      <td>109711</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3012</th>\n",
       "      <td>YHR174W</td>\n",
       "      <td>120322</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2639</th>\n",
       "      <td>YGR192C</td>\n",
       "      <td>173478</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7126</th>\n",
       "      <td>no_feature</td>\n",
       "      <td>414503</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7127</th>\n",
       "      <td>ambiguous</td>\n",
       "      <td>1387186</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7131 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      Gene  Expression\n",
       "7130  alignment_not_unique           0\n",
       "444              YBR121C-A           0\n",
       "6717               YPR177C           0\n",
       "6709             YPR170W-A           0\n",
       "3166               YIL082W           0\n",
       "...                    ...         ...\n",
       "440                YBR118W      109711\n",
       "3012               YHR174W      120322\n",
       "2639               YGR192C      173478\n",
       "7126            no_feature      414503\n",
       "7127             ambiguous     1387186\n",
       "\n",
       "[7131 rows x 2 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.sort_values('Expression')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "df2 = pd.read_csv(\"rawdata/Snf2_rep17_MID89_allLanes_tophat2.0.5.bam.gbgout\",sep=\"\\t\",header=None,names=[\"Gene\",\"Expression\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>Expression</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>15S_rRNA</td>\n",
       "      <td>50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>21S_rRNA</td>\n",
       "      <td>391</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HRA1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ICR1</td>\n",
       "      <td>203</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>LSR1</td>\n",
       "      <td>439</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>NME1</td>\n",
       "      <td>52</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>PWR1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Q0010</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Q0017</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Q0032</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Gene  Expression\n",
       "0  15S_rRNA          50\n",
       "1  21S_rRNA         391\n",
       "2      HRA1           2\n",
       "3      ICR1         203\n",
       "4      LSR1         439\n",
       "5      NME1          52\n",
       "6      PWR1           0\n",
       "7     Q0010           0\n",
       "8     Q0017           0\n",
       "9     Q0032           0"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df2.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>Expression_x</th>\n",
       "      <th>Expression_y</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>15S_rRNA</td>\n",
       "      <td>29</td>\n",
       "      <td>50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>21S_rRNA</td>\n",
       "      <td>221</td>\n",
       "      <td>391</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HRA1</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ICR1</td>\n",
       "      <td>70</td>\n",
       "      <td>203</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>LSR1</td>\n",
       "      <td>314</td>\n",
       "      <td>439</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7126</th>\n",
       "      <td>no_feature</td>\n",
       "      <td>414503</td>\n",
       "      <td>904605</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7127</th>\n",
       "      <td>ambiguous</td>\n",
       "      <td>1387186</td>\n",
       "      <td>1687147</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7128</th>\n",
       "      <td>too_low_aQual</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7129</th>\n",
       "      <td>not_aligned</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7130</th>\n",
       "      <td>alignment_not_unique</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7131 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      Gene  Expression_x  Expression_y\n",
       "0                 15S_rRNA            29            50\n",
       "1                 21S_rRNA           221           391\n",
       "2                     HRA1             0             2\n",
       "3                     ICR1            70           203\n",
       "4                     LSR1           314           439\n",
       "...                    ...           ...           ...\n",
       "7126            no_feature        414503        904605\n",
       "7127             ambiguous       1387186       1687147\n",
       "7128         too_low_aQual             0             0\n",
       "7129           not_aligned             0             0\n",
       "7130  alignment_not_unique             0             0\n",
       "\n",
       "[7131 rows x 3 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.merge(df2,on=\"Gene\",how='outer')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Snf2_rep01_MID96_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep02_MID21_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep03_MID28_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep04_MID11_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep05_MID82_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep06_MID80_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep07_MID26_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep08_MID78_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep09_MID04_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep10_MID38_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep11_MID01_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep12_MID02_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep13_MID59_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep14_MID33_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep15_MID75_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep16_MID72_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep17_MID89_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep18_MID40_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep19_MID84_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep20_MID15_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep21_MID31_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep22_MID53_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep23_MID25_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep24_MID42_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep25_MID81_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep26_MID71_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep27_MID34_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep28_MID74_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep29_MID57_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep30_MID16_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep31_MID77_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep32_MID70_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep33_MID29_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep34_MID14_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep35_MID60_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep36_MID23_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep37_MID46_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep38_MID06_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep39_MID19_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep40_MID12_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep41_MID79_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep42_MID47_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep43_MID73_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep44_MID65_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep45_MID86_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep46_MID35_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep47_MID56_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'Snf2_rep48_MID55_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep01_MID39_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep02_MID50_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep03_MID51_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep04_MID58_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep05_MID93_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep06_MID30_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep07_MID24_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep08_MID20_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep09_MID91_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep10_MID37_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep11_MID54_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep12_MID03_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep13_MID45_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep14_MID52_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep15_MID09_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep16_MID92_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep17_MID66_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep18_MID48_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep19_MID83_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep20_MID90_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep21_MID62_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep22_MID67_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep23_MID36_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep24_MID69_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep25_MID61_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep26_MID08_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep27_MID41_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep28_MID64_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep29_MID68_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep30_MID27_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep31_MID43_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep32_MID95_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep33_MID88_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep34_MID76_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep35_MID32_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep36_MID63_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep37_MID07_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep38_MID87_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep39_MID05_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep40_MID18_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep41_MID13_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep42_MID10_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep43_MID49_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep44_MID22_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep45_MID44_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep46_MID94_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep47_MID17_allLanes_tophat2.0.5.bam.gbgout',\n",
       " 'WT_rep48_MID85_allLanes_tophat2.0.5.bam.gbgout']"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sorted(os.listdir(\"rawdata\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "s = \"WT_rep40_MID18_allLanes_tophat2.0.5.bam.gbgout\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "filesplit = s.split('_')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['WT', 'rep40']"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filesplit[0:2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'WT_rep40'"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"_\".join(filesplit[0:2])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "# \n",
    "# Input data are in individual tab-delimited files, one for each experiment\n",
    "# Here I load each one into its own Pandas dataframe and create a list of frames\n",
    "#\n",
    "\n",
    "frames = []\n",
    "\n",
    "for file in sorted(os.listdir(\"rawdata\")):\n",
    "    if file.endswith(\".gbgout\"):\n",
    "        fs = file.split('_')\n",
    "        name = \"_\".join(fs[0:2])\n",
    "        filepath = os.path.join(\"rawdata/\", file)\n",
    "        frames.append(pd.read_csv(filepath, sep=\"\\t\", header = None, names=['Gene',name]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1\n",
      "2\n",
      "3\n",
      "5\n",
      "7\n",
      "11\n",
      "13\n",
      "17\n"
     ]
    }
   ],
   "source": [
    "l = [1,2,3,5,7,11,13,17]\n",
    "\n",
    "for x in l:\n",
    "    print (x)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = frames[0]\n",
    "for rdf in frames[1:]:\n",
    "    df = df.merge(rdf,on=\"Gene\",how='outer')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>Snf2_rep01</th>\n",
       "      <th>Snf2_rep02</th>\n",
       "      <th>Snf2_rep03</th>\n",
       "      <th>Snf2_rep04</th>\n",
       "      <th>Snf2_rep05</th>\n",
       "      <th>Snf2_rep06</th>\n",
       "      <th>Snf2_rep07</th>\n",
       "      <th>Snf2_rep08</th>\n",
       "      <th>Snf2_rep09</th>\n",
       "      <th>...</th>\n",
       "      <th>WT_rep39</th>\n",
       "      <th>WT_rep40</th>\n",
       "      <th>WT_rep41</th>\n",
       "      <th>WT_rep42</th>\n",
       "      <th>WT_rep43</th>\n",
       "      <th>WT_rep44</th>\n",
       "      <th>WT_rep45</th>\n",
       "      <th>WT_rep46</th>\n",
       "      <th>WT_rep47</th>\n",
       "      <th>WT_rep48</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>15S_rRNA</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>46</td>\n",
       "      <td>3</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>49</td>\n",
       "      <td>9</td>\n",
       "      <td>4</td>\n",
       "      <td>11</td>\n",
       "      <td>12</td>\n",
       "      <td>1</td>\n",
       "      <td>22</td>\n",
       "      <td>12</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>21S_rRNA</td>\n",
       "      <td>31</td>\n",
       "      <td>18</td>\n",
       "      <td>23</td>\n",
       "      <td>44</td>\n",
       "      <td>356</td>\n",
       "      <td>62</td>\n",
       "      <td>35</td>\n",
       "      <td>33</td>\n",
       "      <td>13</td>\n",
       "      <td>...</td>\n",
       "      <td>10</td>\n",
       "      <td>274</td>\n",
       "      <td>49</td>\n",
       "      <td>30</td>\n",
       "      <td>72</td>\n",
       "      <td>58</td>\n",
       "      <td>21</td>\n",
       "      <td>159</td>\n",
       "      <td>107</td>\n",
       "      <td>70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HRA1</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>...</td>\n",
       "      <td>5</td>\n",
       "      <td>3</td>\n",
       "      <td>6</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ICR1</td>\n",
       "      <td>205</td>\n",
       "      <td>196</td>\n",
       "      <td>211</td>\n",
       "      <td>252</td>\n",
       "      <td>127</td>\n",
       "      <td>146</td>\n",
       "      <td>275</td>\n",
       "      <td>160</td>\n",
       "      <td>190</td>\n",
       "      <td>...</td>\n",
       "      <td>85</td>\n",
       "      <td>177</td>\n",
       "      <td>137</td>\n",
       "      <td>118</td>\n",
       "      <td>113</td>\n",
       "      <td>81</td>\n",
       "      <td>142</td>\n",
       "      <td>94</td>\n",
       "      <td>187</td>\n",
       "      <td>106</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>LSR1</td>\n",
       "      <td>210</td>\n",
       "      <td>103</td>\n",
       "      <td>159</td>\n",
       "      <td>260</td>\n",
       "      <td>298</td>\n",
       "      <td>522</td>\n",
       "      <td>303</td>\n",
       "      <td>96</td>\n",
       "      <td>132</td>\n",
       "      <td>...</td>\n",
       "      <td>66</td>\n",
       "      <td>385</td>\n",
       "      <td>232</td>\n",
       "      <td>149</td>\n",
       "      <td>114</td>\n",
       "      <td>81</td>\n",
       "      <td>109</td>\n",
       "      <td>132</td>\n",
       "      <td>243</td>\n",
       "      <td>128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7126</th>\n",
       "      <td>no_feature</td>\n",
       "      <td>888995</td>\n",
       "      <td>641333</td>\n",
       "      <td>544060</td>\n",
       "      <td>955003</td>\n",
       "      <td>616339</td>\n",
       "      <td>764571</td>\n",
       "      <td>1290592</td>\n",
       "      <td>598769</td>\n",
       "      <td>827934</td>\n",
       "      <td>...</td>\n",
       "      <td>403527</td>\n",
       "      <td>1035268</td>\n",
       "      <td>759722</td>\n",
       "      <td>578053</td>\n",
       "      <td>615809</td>\n",
       "      <td>446796</td>\n",
       "      <td>683906</td>\n",
       "      <td>462390</td>\n",
       "      <td>901045</td>\n",
       "      <td>639100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7127</th>\n",
       "      <td>ambiguous</td>\n",
       "      <td>730793</td>\n",
       "      <td>665470</td>\n",
       "      <td>709428</td>\n",
       "      <td>821933</td>\n",
       "      <td>1171511</td>\n",
       "      <td>529946</td>\n",
       "      <td>925353</td>\n",
       "      <td>669715</td>\n",
       "      <td>717502</td>\n",
       "      <td>...</td>\n",
       "      <td>583744</td>\n",
       "      <td>2002465</td>\n",
       "      <td>865662</td>\n",
       "      <td>646805</td>\n",
       "      <td>570375</td>\n",
       "      <td>1011281</td>\n",
       "      <td>899586</td>\n",
       "      <td>596335</td>\n",
       "      <td>1321941</td>\n",
       "      <td>803015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7128</th>\n",
       "      <td>too_low_aQual</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7129</th>\n",
       "      <td>not_aligned</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7130</th>\n",
       "      <td>alignment_not_unique</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7131 rows × 97 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      Gene  Snf2_rep01  Snf2_rep02  Snf2_rep03  Snf2_rep04  \\\n",
       "0                 15S_rRNA           4           2           5           5   \n",
       "1                 21S_rRNA          31          18          23          44   \n",
       "2                     HRA1           5           1           3           1   \n",
       "3                     ICR1         205         196         211         252   \n",
       "4                     LSR1         210         103         159         260   \n",
       "...                    ...         ...         ...         ...         ...   \n",
       "7126            no_feature      888995      641333      544060      955003   \n",
       "7127             ambiguous      730793      665470      709428      821933   \n",
       "7128         too_low_aQual           0           0           0           0   \n",
       "7129           not_aligned           0           0           0           0   \n",
       "7130  alignment_not_unique           0           0           0           0   \n",
       "\n",
       "      Snf2_rep05  Snf2_rep06  Snf2_rep07  Snf2_rep08  Snf2_rep09  ...  \\\n",
       "0             46           3           5           4           2  ...   \n",
       "1            356          62          35          33          13  ...   \n",
       "2              2           1           1           4           4  ...   \n",
       "3            127         146         275         160         190  ...   \n",
       "4            298         522         303          96         132  ...   \n",
       "...          ...         ...         ...         ...         ...  ...   \n",
       "7126      616339      764571     1290592      598769      827934  ...   \n",
       "7127     1171511      529946      925353      669715      717502  ...   \n",
       "7128           0           0           0           0           0  ...   \n",
       "7129           0           0           0           0           0  ...   \n",
       "7130           0           0           0           0           0  ...   \n",
       "\n",
       "      WT_rep39  WT_rep40  WT_rep41  WT_rep42  WT_rep43  WT_rep44  WT_rep45  \\\n",
       "0            0        49         9         4        11        12         1   \n",
       "1           10       274        49        30        72        58        21   \n",
       "2            5         3         6         5         2         2         2   \n",
       "3           85       177       137       118       113        81       142   \n",
       "4           66       385       232       149       114        81       109   \n",
       "...        ...       ...       ...       ...       ...       ...       ...   \n",
       "7126    403527   1035268    759722    578053    615809    446796    683906   \n",
       "7127    583744   2002465    865662    646805    570375   1011281    899586   \n",
       "7128         0         0         0         0         0         0         0   \n",
       "7129         0         0         0         0         0         0         0   \n",
       "7130         0         0         0         0         0         0         0   \n",
       "\n",
       "      WT_rep46  WT_rep47  WT_rep48  \n",
       "0           22        12         4  \n",
       "1          159       107        70  \n",
       "2            5         2         1  \n",
       "3           94       187       106  \n",
       "4          132       243       128  \n",
       "...        ...       ...       ...  \n",
       "7126    462390    901045    639100  \n",
       "7127    596335   1321941    803015  \n",
       "7128         0         0         0  \n",
       "7129         0         0         0  \n",
       "7130         0         0         0  \n",
       "\n",
       "[7131 rows x 97 columns]"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>WT_rep48</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>15S_rRNA</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>21S_rRNA</td>\n",
       "      <td>70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HRA1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ICR1</td>\n",
       "      <td>106</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>LSR1</td>\n",
       "      <td>128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7126</th>\n",
       "      <td>no_feature</td>\n",
       "      <td>639100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7127</th>\n",
       "      <td>ambiguous</td>\n",
       "      <td>803015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7128</th>\n",
       "      <td>too_low_aQual</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7129</th>\n",
       "      <td>not_aligned</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7130</th>\n",
       "      <td>alignment_not_unique</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7131 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      Gene  WT_rep48\n",
       "0                 15S_rRNA         4\n",
       "1                 21S_rRNA        70\n",
       "2                     HRA1         1\n",
       "3                     ICR1       106\n",
       "4                     LSR1       128\n",
       "...                    ...       ...\n",
       "7126            no_feature    639100\n",
       "7127             ambiguous    803015\n",
       "7128         too_low_aQual         0\n",
       "7129           not_aligned         0\n",
       "7130  alignment_not_unique         0\n",
       "\n",
       "[7131 rows x 2 columns]"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "frames[-1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#\n",
    "# This combines all of the individual frames into one dataframe, combining on the \"Gene\" column\n",
    "#\n",
    "\n",
    "df = reduce(lambda  left,right: pd.merge(left,right,on=['Gene'], how='outer'), frames)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "#\n",
    "# Sort columns by name\n",
    "#\n",
    "\n",
    "df = df.reindex(sorted(df.columns),axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>Snf2_rep01</th>\n",
       "      <th>Snf2_rep02</th>\n",
       "      <th>Snf2_rep03</th>\n",
       "      <th>Snf2_rep04</th>\n",
       "      <th>Snf2_rep05</th>\n",
       "      <th>Snf2_rep06</th>\n",
       "      <th>Snf2_rep07</th>\n",
       "      <th>Snf2_rep08</th>\n",
       "      <th>Snf2_rep09</th>\n",
       "      <th>...</th>\n",
       "      <th>WT_rep39</th>\n",
       "      <th>WT_rep40</th>\n",
       "      <th>WT_rep41</th>\n",
       "      <th>WT_rep42</th>\n",
       "      <th>WT_rep43</th>\n",
       "      <th>WT_rep44</th>\n",
       "      <th>WT_rep45</th>\n",
       "      <th>WT_rep46</th>\n",
       "      <th>WT_rep47</th>\n",
       "      <th>WT_rep48</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>15S_rRNA</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>46</td>\n",
       "      <td>3</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>49</td>\n",
       "      <td>9</td>\n",
       "      <td>4</td>\n",
       "      <td>11</td>\n",
       "      <td>12</td>\n",
       "      <td>1</td>\n",
       "      <td>22</td>\n",
       "      <td>12</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>21S_rRNA</td>\n",
       "      <td>31</td>\n",
       "      <td>18</td>\n",
       "      <td>23</td>\n",
       "      <td>44</td>\n",
       "      <td>356</td>\n",
       "      <td>62</td>\n",
       "      <td>35</td>\n",
       "      <td>33</td>\n",
       "      <td>13</td>\n",
       "      <td>...</td>\n",
       "      <td>10</td>\n",
       "      <td>274</td>\n",
       "      <td>49</td>\n",
       "      <td>30</td>\n",
       "      <td>72</td>\n",
       "      <td>58</td>\n",
       "      <td>21</td>\n",
       "      <td>159</td>\n",
       "      <td>107</td>\n",
       "      <td>70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HRA1</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>...</td>\n",
       "      <td>5</td>\n",
       "      <td>3</td>\n",
       "      <td>6</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ICR1</td>\n",
       "      <td>205</td>\n",
       "      <td>196</td>\n",
       "      <td>211</td>\n",
       "      <td>252</td>\n",
       "      <td>127</td>\n",
       "      <td>146</td>\n",
       "      <td>275</td>\n",
       "      <td>160</td>\n",
       "      <td>190</td>\n",
       "      <td>...</td>\n",
       "      <td>85</td>\n",
       "      <td>177</td>\n",
       "      <td>137</td>\n",
       "      <td>118</td>\n",
       "      <td>113</td>\n",
       "      <td>81</td>\n",
       "      <td>142</td>\n",
       "      <td>94</td>\n",
       "      <td>187</td>\n",
       "      <td>106</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>LSR1</td>\n",
       "      <td>210</td>\n",
       "      <td>103</td>\n",
       "      <td>159</td>\n",
       "      <td>260</td>\n",
       "      <td>298</td>\n",
       "      <td>522</td>\n",
       "      <td>303</td>\n",
       "      <td>96</td>\n",
       "      <td>132</td>\n",
       "      <td>...</td>\n",
       "      <td>66</td>\n",
       "      <td>385</td>\n",
       "      <td>232</td>\n",
       "      <td>149</td>\n",
       "      <td>114</td>\n",
       "      <td>81</td>\n",
       "      <td>109</td>\n",
       "      <td>132</td>\n",
       "      <td>243</td>\n",
       "      <td>128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7126</th>\n",
       "      <td>no_feature</td>\n",
       "      <td>888995</td>\n",
       "      <td>641333</td>\n",
       "      <td>544060</td>\n",
       "      <td>955003</td>\n",
       "      <td>616339</td>\n",
       "      <td>764571</td>\n",
       "      <td>1290592</td>\n",
       "      <td>598769</td>\n",
       "      <td>827934</td>\n",
       "      <td>...</td>\n",
       "      <td>403527</td>\n",
       "      <td>1035268</td>\n",
       "      <td>759722</td>\n",
       "      <td>578053</td>\n",
       "      <td>615809</td>\n",
       "      <td>446796</td>\n",
       "      <td>683906</td>\n",
       "      <td>462390</td>\n",
       "      <td>901045</td>\n",
       "      <td>639100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7127</th>\n",
       "      <td>ambiguous</td>\n",
       "      <td>730793</td>\n",
       "      <td>665470</td>\n",
       "      <td>709428</td>\n",
       "      <td>821933</td>\n",
       "      <td>1171511</td>\n",
       "      <td>529946</td>\n",
       "      <td>925353</td>\n",
       "      <td>669715</td>\n",
       "      <td>717502</td>\n",
       "      <td>...</td>\n",
       "      <td>583744</td>\n",
       "      <td>2002465</td>\n",
       "      <td>865662</td>\n",
       "      <td>646805</td>\n",
       "      <td>570375</td>\n",
       "      <td>1011281</td>\n",
       "      <td>899586</td>\n",
       "      <td>596335</td>\n",
       "      <td>1321941</td>\n",
       "      <td>803015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7128</th>\n",
       "      <td>too_low_aQual</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7129</th>\n",
       "      <td>not_aligned</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7130</th>\n",
       "      <td>alignment_not_unique</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7131 rows × 97 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      Gene  Snf2_rep01  Snf2_rep02  Snf2_rep03  Snf2_rep04  \\\n",
       "0                 15S_rRNA           4           2           5           5   \n",
       "1                 21S_rRNA          31          18          23          44   \n",
       "2                     HRA1           5           1           3           1   \n",
       "3                     ICR1         205         196         211         252   \n",
       "4                     LSR1         210         103         159         260   \n",
       "...                    ...         ...         ...         ...         ...   \n",
       "7126            no_feature      888995      641333      544060      955003   \n",
       "7127             ambiguous      730793      665470      709428      821933   \n",
       "7128         too_low_aQual           0           0           0           0   \n",
       "7129           not_aligned           0           0           0           0   \n",
       "7130  alignment_not_unique           0           0           0           0   \n",
       "\n",
       "      Snf2_rep05  Snf2_rep06  Snf2_rep07  Snf2_rep08  Snf2_rep09  ...  \\\n",
       "0             46           3           5           4           2  ...   \n",
       "1            356          62          35          33          13  ...   \n",
       "2              2           1           1           4           4  ...   \n",
       "3            127         146         275         160         190  ...   \n",
       "4            298         522         303          96         132  ...   \n",
       "...          ...         ...         ...         ...         ...  ...   \n",
       "7126      616339      764571     1290592      598769      827934  ...   \n",
       "7127     1171511      529946      925353      669715      717502  ...   \n",
       "7128           0           0           0           0           0  ...   \n",
       "7129           0           0           0           0           0  ...   \n",
       "7130           0           0           0           0           0  ...   \n",
       "\n",
       "      WT_rep39  WT_rep40  WT_rep41  WT_rep42  WT_rep43  WT_rep44  WT_rep45  \\\n",
       "0            0        49         9         4        11        12         1   \n",
       "1           10       274        49        30        72        58        21   \n",
       "2            5         3         6         5         2         2         2   \n",
       "3           85       177       137       118       113        81       142   \n",
       "4           66       385       232       149       114        81       109   \n",
       "...        ...       ...       ...       ...       ...       ...       ...   \n",
       "7126    403527   1035268    759722    578053    615809    446796    683906   \n",
       "7127    583744   2002465    865662    646805    570375   1011281    899586   \n",
       "7128         0         0         0         0         0         0         0   \n",
       "7129         0         0         0         0         0         0         0   \n",
       "7130         0         0         0         0         0         0         0   \n",
       "\n",
       "      WT_rep46  WT_rep47  WT_rep48  \n",
       "0           22        12         4  \n",
       "1          159       107        70  \n",
       "2            5         2         1  \n",
       "3           94       187       106  \n",
       "4          132       243       128  \n",
       "...        ...       ...       ...  \n",
       "7126    462390    901045    639100  \n",
       "7127    596335   1321941    803015  \n",
       "7128         0         0         0  \n",
       "7129         0         0         0  \n",
       "7130         0         0         0  \n",
       "\n",
       "[7131 rows x 97 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0       False\n",
       "1       False\n",
       "2       False\n",
       "3       False\n",
       "4       False\n",
       "        ...  \n",
       "7126    False\n",
       "7127    False\n",
       "7128    False\n",
       "7129    False\n",
       "7130    False\n",
       "Name: Gene, Length: 7131, dtype: bool"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['Gene'].str.startswith('Y')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>Snf2_rep01</th>\n",
       "      <th>Snf2_rep02</th>\n",
       "      <th>Snf2_rep03</th>\n",
       "      <th>Snf2_rep04</th>\n",
       "      <th>Snf2_rep05</th>\n",
       "      <th>Snf2_rep06</th>\n",
       "      <th>Snf2_rep07</th>\n",
       "      <th>Snf2_rep08</th>\n",
       "      <th>Snf2_rep09</th>\n",
       "      <th>...</th>\n",
       "      <th>WT_rep39</th>\n",
       "      <th>WT_rep40</th>\n",
       "      <th>WT_rep41</th>\n",
       "      <th>WT_rep42</th>\n",
       "      <th>WT_rep43</th>\n",
       "      <th>WT_rep44</th>\n",
       "      <th>WT_rep45</th>\n",
       "      <th>WT_rep46</th>\n",
       "      <th>WT_rep47</th>\n",
       "      <th>WT_rep48</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>TLC1</td>\n",
       "      <td>2314</td>\n",
       "      <td>1587</td>\n",
       "      <td>1926</td>\n",
       "      <td>2159</td>\n",
       "      <td>1537</td>\n",
       "      <td>920</td>\n",
       "      <td>3054</td>\n",
       "      <td>1476</td>\n",
       "      <td>1935</td>\n",
       "      <td>...</td>\n",
       "      <td>872</td>\n",
       "      <td>1802</td>\n",
       "      <td>1564</td>\n",
       "      <td>1144</td>\n",
       "      <td>1264</td>\n",
       "      <td>698</td>\n",
       "      <td>1577</td>\n",
       "      <td>978</td>\n",
       "      <td>1823</td>\n",
       "      <td>1297</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>YAL003W</td>\n",
       "      <td>7296</td>\n",
       "      <td>6129</td>\n",
       "      <td>6464</td>\n",
       "      <td>7278</td>\n",
       "      <td>5613</td>\n",
       "      <td>3046</td>\n",
       "      <td>7687</td>\n",
       "      <td>6714</td>\n",
       "      <td>8382</td>\n",
       "      <td>...</td>\n",
       "      <td>9291</td>\n",
       "      <td>12055</td>\n",
       "      <td>13542</td>\n",
       "      <td>8441</td>\n",
       "      <td>7337</td>\n",
       "      <td>8639</td>\n",
       "      <td>14315</td>\n",
       "      <td>6618</td>\n",
       "      <td>14745</td>\n",
       "      <td>9878</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>YAL005C</td>\n",
       "      <td>9851</td>\n",
       "      <td>10226</td>\n",
       "      <td>12006</td>\n",
       "      <td>10714</td>\n",
       "      <td>8575</td>\n",
       "      <td>2791</td>\n",
       "      <td>11764</td>\n",
       "      <td>8237</td>\n",
       "      <td>12213</td>\n",
       "      <td>...</td>\n",
       "      <td>10177</td>\n",
       "      <td>21176</td>\n",
       "      <td>13270</td>\n",
       "      <td>12337</td>\n",
       "      <td>14303</td>\n",
       "      <td>14268</td>\n",
       "      <td>14993</td>\n",
       "      <td>9607</td>\n",
       "      <td>16075</td>\n",
       "      <td>14797</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>YAL007C</td>\n",
       "      <td>1519</td>\n",
       "      <td>1276</td>\n",
       "      <td>1364</td>\n",
       "      <td>1611</td>\n",
       "      <td>1116</td>\n",
       "      <td>591</td>\n",
       "      <td>1951</td>\n",
       "      <td>1002</td>\n",
       "      <td>1665</td>\n",
       "      <td>...</td>\n",
       "      <td>753</td>\n",
       "      <td>1399</td>\n",
       "      <td>1095</td>\n",
       "      <td>818</td>\n",
       "      <td>755</td>\n",
       "      <td>753</td>\n",
       "      <td>1137</td>\n",
       "      <td>695</td>\n",
       "      <td>1346</td>\n",
       "      <td>910</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>YAL012W</td>\n",
       "      <td>7277</td>\n",
       "      <td>7124</td>\n",
       "      <td>7545</td>\n",
       "      <td>8022</td>\n",
       "      <td>5875</td>\n",
       "      <td>7223</td>\n",
       "      <td>8985</td>\n",
       "      <td>6530</td>\n",
       "      <td>8068</td>\n",
       "      <td>...</td>\n",
       "      <td>4421</td>\n",
       "      <td>6788</td>\n",
       "      <td>5605</td>\n",
       "      <td>4121</td>\n",
       "      <td>3639</td>\n",
       "      <td>5998</td>\n",
       "      <td>7428</td>\n",
       "      <td>2768</td>\n",
       "      <td>7148</td>\n",
       "      <td>4273</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6738</th>\n",
       "      <td>YPR198W</td>\n",
       "      <td>1009</td>\n",
       "      <td>879</td>\n",
       "      <td>833</td>\n",
       "      <td>1116</td>\n",
       "      <td>749</td>\n",
       "      <td>761</td>\n",
       "      <td>1355</td>\n",
       "      <td>649</td>\n",
       "      <td>1117</td>\n",
       "      <td>...</td>\n",
       "      <td>416</td>\n",
       "      <td>850</td>\n",
       "      <td>568</td>\n",
       "      <td>448</td>\n",
       "      <td>569</td>\n",
       "      <td>399</td>\n",
       "      <td>663</td>\n",
       "      <td>349</td>\n",
       "      <td>691</td>\n",
       "      <td>569</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6745</th>\n",
       "      <td>YPR204W</td>\n",
       "      <td>1849</td>\n",
       "      <td>1544</td>\n",
       "      <td>1611</td>\n",
       "      <td>1834</td>\n",
       "      <td>1138</td>\n",
       "      <td>2344</td>\n",
       "      <td>2367</td>\n",
       "      <td>1619</td>\n",
       "      <td>1481</td>\n",
       "      <td>...</td>\n",
       "      <td>1010</td>\n",
       "      <td>1419</td>\n",
       "      <td>1005</td>\n",
       "      <td>778</td>\n",
       "      <td>596</td>\n",
       "      <td>1448</td>\n",
       "      <td>1602</td>\n",
       "      <td>597</td>\n",
       "      <td>1393</td>\n",
       "      <td>1010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6824</th>\n",
       "      <td>snR86</td>\n",
       "      <td>1328</td>\n",
       "      <td>897</td>\n",
       "      <td>1063</td>\n",
       "      <td>1265</td>\n",
       "      <td>935</td>\n",
       "      <td>1498</td>\n",
       "      <td>1757</td>\n",
       "      <td>823</td>\n",
       "      <td>869</td>\n",
       "      <td>...</td>\n",
       "      <td>424</td>\n",
       "      <td>1125</td>\n",
       "      <td>776</td>\n",
       "      <td>593</td>\n",
       "      <td>881</td>\n",
       "      <td>437</td>\n",
       "      <td>796</td>\n",
       "      <td>653</td>\n",
       "      <td>1045</td>\n",
       "      <td>819</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7126</th>\n",
       "      <td>no_feature</td>\n",
       "      <td>888995</td>\n",
       "      <td>641333</td>\n",
       "      <td>544060</td>\n",
       "      <td>955003</td>\n",
       "      <td>616339</td>\n",
       "      <td>764571</td>\n",
       "      <td>1290592</td>\n",
       "      <td>598769</td>\n",
       "      <td>827934</td>\n",
       "      <td>...</td>\n",
       "      <td>403527</td>\n",
       "      <td>1035268</td>\n",
       "      <td>759722</td>\n",
       "      <td>578053</td>\n",
       "      <td>615809</td>\n",
       "      <td>446796</td>\n",
       "      <td>683906</td>\n",
       "      <td>462390</td>\n",
       "      <td>901045</td>\n",
       "      <td>639100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7127</th>\n",
       "      <td>ambiguous</td>\n",
       "      <td>730793</td>\n",
       "      <td>665470</td>\n",
       "      <td>709428</td>\n",
       "      <td>821933</td>\n",
       "      <td>1171511</td>\n",
       "      <td>529946</td>\n",
       "      <td>925353</td>\n",
       "      <td>669715</td>\n",
       "      <td>717502</td>\n",
       "      <td>...</td>\n",
       "      <td>583744</td>\n",
       "      <td>2002465</td>\n",
       "      <td>865662</td>\n",
       "      <td>646805</td>\n",
       "      <td>570375</td>\n",
       "      <td>1011281</td>\n",
       "      <td>899586</td>\n",
       "      <td>596335</td>\n",
       "      <td>1321941</td>\n",
       "      <td>803015</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2143 rows × 97 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            Gene  Snf2_rep01  Snf2_rep02  Snf2_rep03  Snf2_rep04  Snf2_rep05  \\\n",
       "60          TLC1        2314        1587        1926        2159        1537   \n",
       "63       YAL003W        7296        6129        6464        7278        5613   \n",
       "65       YAL005C        9851       10226       12006       10714        8575   \n",
       "66       YAL007C        1519        1276        1364        1611        1116   \n",
       "71       YAL012W        7277        7124        7545        8022        5875   \n",
       "...          ...         ...         ...         ...         ...         ...   \n",
       "6738     YPR198W        1009         879         833        1116         749   \n",
       "6745     YPR204W        1849        1544        1611        1834        1138   \n",
       "6824       snR86        1328         897        1063        1265         935   \n",
       "7126  no_feature      888995      641333      544060      955003      616339   \n",
       "7127   ambiguous      730793      665470      709428      821933     1171511   \n",
       "\n",
       "      Snf2_rep06  Snf2_rep07  Snf2_rep08  Snf2_rep09  ...  WT_rep39  WT_rep40  \\\n",
       "60           920        3054        1476        1935  ...       872      1802   \n",
       "63          3046        7687        6714        8382  ...      9291     12055   \n",
       "65          2791       11764        8237       12213  ...     10177     21176   \n",
       "66           591        1951        1002        1665  ...       753      1399   \n",
       "71          7223        8985        6530        8068  ...      4421      6788   \n",
       "...          ...         ...         ...         ...  ...       ...       ...   \n",
       "6738         761        1355         649        1117  ...       416       850   \n",
       "6745        2344        2367        1619        1481  ...      1010      1419   \n",
       "6824        1498        1757         823         869  ...       424      1125   \n",
       "7126      764571     1290592      598769      827934  ...    403527   1035268   \n",
       "7127      529946      925353      669715      717502  ...    583744   2002465   \n",
       "\n",
       "      WT_rep41  WT_rep42  WT_rep43  WT_rep44  WT_rep45  WT_rep46  WT_rep47  \\\n",
       "60        1564      1144      1264       698      1577       978      1823   \n",
       "63       13542      8441      7337      8639     14315      6618     14745   \n",
       "65       13270     12337     14303     14268     14993      9607     16075   \n",
       "66        1095       818       755       753      1137       695      1346   \n",
       "71        5605      4121      3639      5998      7428      2768      7148   \n",
       "...        ...       ...       ...       ...       ...       ...       ...   \n",
       "6738       568       448       569       399       663       349       691   \n",
       "6745      1005       778       596      1448      1602       597      1393   \n",
       "6824       776       593       881       437       796       653      1045   \n",
       "7126    759722    578053    615809    446796    683906    462390    901045   \n",
       "7127    865662    646805    570375   1011281    899586    596335   1321941   \n",
       "\n",
       "      WT_rep48  \n",
       "60        1297  \n",
       "63        9878  \n",
       "65       14797  \n",
       "66         910  \n",
       "71        4273  \n",
       "...        ...  \n",
       "6738       569  \n",
       "6745      1010  \n",
       "6824       819  \n",
       "7126    639100  \n",
       "7127    803015  \n",
       "\n",
       "[2143 rows x 97 columns]"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[df['Snf2_rep01'] > 1000]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "df = df[df['Gene'].str.startswith('Y')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [],
   "source": [
    "#\n",
    "# Use only genes starting with \"Y\" which are protein-coding genes\n",
    "#\n",
    "\n",
    "df = df[df['Gene'].str.startswith('Y')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [],
   "source": [
    "#\n",
    "# Index dataframe on 'Gene'\n",
    "#\n",
    "\n",
    "df = df.set_index('Gene')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Save\n",
    "\n",
    "df.to_csv(\"Barton_combined_Ygenes.txt\", sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.to_pickle(\"Barton_combined_Ygenes.pkl\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
